```
Error running mint automation
+ mkdir -p 7949-3e5e1db/gopath/src/github.com/minio
+ git -C 7949-3e5e1db/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 7949-3e5e1db/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 7949-3e5e1db/gopath/src/github.com/minio/minio fetch --quiet minio pull/7949/head:pr7949
+ git -C 7949-3e5e1db/gopath/src/github.com/minio/minio checkout --quiet pr7949
+ GO111MODULE=on
+ GOPATH=/mnt/nvme/mint-auto/7949-3e5e1db/gopath
+ make -C 7949-3e5e1db/gopath/src/github.com/minio/minio --quiet
go: finding github.com/gorilla/handlers v1.4.0
go: finding github.com/Azure/go-autorest v11.7.0+incompatible
go: finding github.com/modern-go/reflect2 v1.0.1
go: finding github.com/baiyubin/aliyun-sts-go-sdk v0.0.0-20180326062324-cfa1a18b161f
go: finding github.com/tidwall/gjson v1.2.1
go: finding github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: finding github.com/minio/parquet-go v0.0.0-20190318185229-9d767baf1679
go: finding github.com/rs/cors v1.6.0
go: finding google.golang.org/api v0.4.0
go: finding gopkg.in/Shopify/sarama.v1 v1.20.0
go: finding github.com/minio/lsync v1.0.1
go: finding github.com/gorilla/mux v1.7.0
go: finding github.com/dnaeon/go-vcr v1.0.1
go: finding github.com/smartystreets/goconvey v0.0.0-20190222223459-a17d461953aa
go: finding github.com/cheggaaa/pb v1.0.28
go: finding github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: finding github.com/coreos/etcd v3.3.12+incompatible
go: finding github.com/aws/aws-sdk-go v1.20.21
go: finding github.com/nats-io/stan.go v0.4.5
go: finding github.com/hashicorp/golang-lru v0.5.1
go: finding github.com/klauspost/readahead v1.3.0
go: finding go.uber.org/atomic v1.3.2
go: finding golang.org/x/net v0.0.0-20190301231341-16b79f2e4e95
go: finding github.com/grpc-ecosystem/grpc-gateway v1.9.0
go: finding github.com/rjeczalik/notify v0.9.2
go: finding github.com/minio/minio-go v0.0.0-20190327203652-5325257a208f
go: finding github.com/hashicorp/vault v1.1.0
go: finding github.com/dustin/go-humanize v1.0.0
go: finding github.com/minio/sio v0.2.0
go: finding golang.org/x/crypto v0.0.0-20190621222207-cc06ce4a13d4
go: finding gopkg.in/yaml.v2 v2.2.2
go: finding github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: finding golang.org/x/crypto v0.0.0-20181106171534-e4dc69e5b2fd
go: finding github.com/pierrec/lz4 v2.0.5+incompatible
go: finding github.com/ghodss/yaml v1.0.0
go: finding github.com/rogpeppe/fastuuid v0.0.0-20150106093220-6724a57986af
go: finding github.com/nats-io/nats-server v1.4.1
go: finding golang.org/x/sync v0.0.0-20190227155943-e225da77a7e6
go: finding github.com/Azure/azure-sdk-for-go v27.0.0+incompatible
go: finding github.com/tidwall/pretty v1.0.0
go: finding google.golang.org/grpc v1.19.0
go: finding golang.org/x/sys v0.0.0-20180926160741-c2ed4eda69e7
go: finding github.com/minio/cli v1.21.0
go: finding github.com/klauspost/pgzip v1.2.1
go: finding github.com/eclipse/paho.mqtt.golang v1.1.2-0.20190322152051-20337d8c3947
go: finding github.com/skyrings/skyring-common v0.0.0-20160929130248-d1c0bb1cbd5e
go: finding github.com/mitchellh/go-homedir v1.1.0
go: finding github.com/mattn/go-isatty v0.0.7
go: finding golang.org/x/sys v0.0.0-20190626221950-04f50cda93cb
go: finding github.com/klauspost/reedsolomon v1.9.1
go: finding github.com/kurin/blazer v0.5.4-0.20190613185654-cf2f27cc0be3
go: finding golang.org/x/sys v0.0.0-20180830151530-49385e6e1522
go: finding github.com/smartystreets/assertions v0.0.0-20190215210624-980c5ac6f3ac
go: finding github.com/golang/protobuf v1.2.0
go: finding gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7
go: finding contrib.go.opencensus.io/exporter/ocagent v0.5.0
go: finding github.com/golang/protobuf v1.3.1
go: finding gopkg.in/ini.v1 v1.41.0
go: finding github.com/BurntSushi/toml v0.3.1
go: finding golang.org/x/crypto v0.0.0-20190228161510-8dd112bcdc25
go: finding github.com/tidwall/sjson v1.0.4
go: finding go.opencensus.io v0.21.0
go: finding golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
go: finding github.com/minio/minio-go/v6 v6.0.29
go: finding github.com/minio/mc v0.0.0-20190529152718-f4bb0b8850cb
go: finding honnef.co/go/tools v0.0.0-20190102054323-c2f93a96b099
go: finding golang.org/x/oauth2 v0.0.0-20180821212333-d2e6202438be
go: finding gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
go: finding golang.org/x/net v0.0.0-20190620200207-3b0461eec859
go: finding github.com/nats-io/go-nats-streaming v0.4.4
go: finding github.com/satori/go.uuid v1.2.0
go: finding github.com/census-instrumentation/opencensus-proto v0.2.0
go: finding golang.org/x/crypto v0.0.0-20190128193316-c7b33c32a30b
go: finding github.com/hashicorp/go-rootcerts v1.0.1
go: finding github.com/ncw/directio v1.0.5
go: finding cloud.google.com/go v0.26.0
go: finding github.com/minio/minio v0.0.0-20190510004154-ac3b59645e92
go: finding github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: finding github.com/mattn/go-colorable v0.1.1
go: finding github.com/ryanuber/go-glob v1.0.0
go: finding github.com/minio/minio-go v0.0.0-20190227180923-59af836a7e6d
go: finding github.com/nats-io/gnatsd v1.4.1
go: finding gopkg.in/olivere/elastic.v5 v5.0.80
go: finding golang.org/x/text v0.3.2
go: finding golang.org/x/lint v0.0.0-20181026193005-c67002cb31c3
go: finding github.com/jtolds/gls v4.20.0+incompatible
go: finding cloud.google.com/go v0.37.2
go: finding golang.org/x/lint v0.0.0-20190301231843-5614ed5bae6f
go: finding github.com/nats-io/nats-streaming-server v0.14.2
go: finding github.com/client9/misspell v0.3.4
go: finding github.com/smartystreets/assertions v0.0.0-20190116191733-b6c0e53d7304
go: finding github.com/grpc-ecosystem/grpc-gateway v1.8.5
go: finding golang.org/x/crypto v0.0.0-20190513172903-22d7a77e9e5f
go: finding github.com/DataDog/zstd v1.4.0
go: finding github.com/hashicorp/go-version v1.1.0
go: finding github.com/aliyun/aliyun-oss-go-sdk v0.0.0-20190307165228-86c17b95fcd5
go: finding golang.org/x/tools v0.0.0-20190114222345-bf090417da8b
go: finding golang.org/x/net v0.0.0-20190213061140-3a22650c66bd
go: finding gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127
go: finding github.com/minio/minio-go/v6 v6.0.26
go: finding github.com/lib/pq v1.0.0
go: finding golang.org/x/text v0.3.0
go: finding github.com/streadway/amqp v0.0.0-20190402114354-16ed540749f6
go: finding github.com/gomodule/redigo v2.0.0+incompatible
go: finding golang.org/x/tools v0.0.0-20190312170243-e65039ee4138
go: finding github.com/google/go-cmp v0.2.0
go: finding github.com/pkg/profile v1.3.0
go: finding github.com/gopherjs/gopherjs v0.0.0-20181103185306-d547d1d9531e
go: finding golang.org/x/sys v0.0.0-20190412213103-97732733099d
go: finding github.com/smartystreets/goconvey v0.0.0-20181108003508-044398e4856c
go: finding golang.org/x/net v0.0.0-20190522155817-f3200d17e092
go: finding github.com/smartystreets/assertions v0.0.0-20180927180507-b2de0cb4f26d
go: finding github.com/coredns/coredns v1.4.0
go: finding golang.org/x/time v0.0.0-20181108054448-85acf8d2951c
go: finding github.com/google/martian v2.1.0+incompatible
go: finding github.com/jstemmer/go-junit-report v0.0.0-20190106144839-af01ea7f8024
go: finding google.golang.org/genproto v0.0.0-20190307195333-5fe7a883aa19
go: finding golang.org/x/sys v0.0.0-20190222072716-a9d3bda3a223
go: finding github.com/klauspost/cpuid v1.2.1
go: finding github.com/google/btree v0.0.0-20180813153112-4030bb1f1f0c
go: finding github.com/tidwall/match v1.0.1
go: finding gopkg.in/h2non/filetype.v1 v1.0.5
go: finding github.com/jtolds/gls v4.2.1+incompatible
go: finding github.com/hashicorp/go-sockaddr v1.0.2
go: finding golang.org/x/text v0.3.1-0.20180807135948-17ff2d5776d2
go: finding go.opencensus.io v0.19.2
go: finding go.etcd.io/bbolt v1.3.3
go: finding github.com/golang/groupcache v0.0.0-20190702054246-869f871628b6
go: finding github.com/ugorji/go v1.1.4
go: finding github.com/smartystreets/goconvey v0.0.0-20190330032615-68dc04aab96a
go: finding github.com/eapache/go-resiliency v1.2.0
go: finding gopkg.in/resty.v1 v1.12.0
go: finding github.com/ryanuber/columnize v2.1.0+incompatible
go: finding github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: finding github.com/pkg/xattr v0.4.1
go: finding google.golang.org/grpc v1.20.1
go: finding github.com/golang/snappy v0.0.1
go: finding golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a
go: finding github.com/go-sql-driver/mysql v1.4.1
go: finding google.golang.org/appengine v1.4.0
go: finding golang.org/x/net v0.0.0-20180826012351-8a410e7b638d
go: finding github.com/minio/cli v1.20.0
go: finding github.com/minio/dsync/v2 v2.0.0
go: finding golang.org/x/exp v0.0.0-20190121172915-509febef88a4
go: finding golang.org/x/lint v0.0.0-20190313153728-d0100b6bd8b3
go: finding github.com/fatih/structs v1.1.0
go: finding golang.org/x/build v0.0.0-20190314133821-5284462c4bec
go: finding github.com/google/pprof v0.0.0-20181206194817-3ea8567a2e57
go: finding github.com/nsqio/go-nsq v1.0.7
go: finding github.com/mattn/go-isatty v0.0.5
go: finding github.com/miekg/dns v1.1.8
go: finding github.com/smartystreets/go-aws-auth v0.0.0-20180515143844-0c1422d1fdb9
go: finding github.com/hashicorp/go-uuid v1.0.1
go: finding google.golang.org/appengine v1.1.0
go: finding golang.org/x/tools v0.0.0-20190311212946-11955173bddd
go: finding github.com/pkg/errors v0.8.0
go: finding golang.org/x/sys v0.0.0-20181107165924-66b7b1311ac8
go: finding gopkg.in/ini.v1 v1.42.0
go: finding github.com/fatih/color v1.7.0
go: finding golang.org/x/sys v0.0.0-20190124100055-b90733256f2e
go: finding github.com/hashicorp/golang-lru v0.5.0
go: finding golang.org/x/sys v0.0.0-20181021155630-eda9bb28ed51
go: finding github.com/nats-io/nats.go v1.8.0
go: finding github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: finding github.com/marstr/guid v1.1.0
go: finding github.com/a8m/mark v0.1.1-0.20170507133748-44f2db618845
go: finding github.com/coreos/go-semver v0.3.0
go: finding github.com/posener/complete v1.2.1
go: finding cloud.google.com/go v0.31.0
go: finding go.opencensus.io v0.18.0
go: finding github.com/minio/highwayhash v1.0.0
go: finding github.com/minio/sha256-simd v0.0.0-20190328051042-05b4dd3047e5
go: finding google.golang.org/api v0.0.0-20181030000543-1d582fd0359e
go: finding github.com/hashicorp/raft v1.1.0
go: finding git.apache.org/thrift.git v0.12.0
go: finding github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: finding github.com/matttproud/golang_protobuf_extensions v1.0.1
go: finding github.com/gopherjs/gopherjs v0.0.0-20181017120253-0766667cb4d1
go: finding golang.org/x/sys v0.0.0-20190304154630-e844e0132e93
go: finding golang.org/x/sys v0.0.0-20180909124046-d0be0721c37e
go: finding golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f
go: finding golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
go: finding github.com/jellevandenhooff/dkim v0.0.0-20150330215556-f50fe3d243e1
go: finding golang.org/x/net v0.0.0-20181220203305-927f97764cc3
go: finding github.com/nats-io/nuid v1.0.1
go: finding go.uber.org/zap v1.10.0
go: finding github.com/google/go-querystring v1.0.0
go: finding github.com/elazarl/go-bindata-assetfs v1.0.0
go: finding github.com/kr/pty v1.1.3
go: finding github.com/alecthomas/participle v0.2.1
go: finding github.com/hashicorp/errwrap v1.0.0
go: finding github.com/sirupsen/logrus v1.4.2
go: finding github.com/hashicorp/hcl v1.0.0
go: finding github.com/djherbis/atime v1.0.0
go: finding golang.org/x/crypto v0.0.0-20181030102418-4d3f4d9ffa16
go: finding golang.org/x/tools v0.0.0-20180917221912-90fa682c2a6e
go: finding github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
go: finding golang.org/x/net v0.0.0-20180906233101-161cd47e91fd
go: finding github.com/prometheus/common v0.0.0-20180801064454-c7de2306084e
go: finding gopkg.in/urfave/cli.v1 v1.20.0
go: finding github.com/coreos/bbolt v1.3.3
go: finding github.com/stretchr/testify v1.2.2
go: finding github.com/prometheus/client_model v0.0.0-20180712105110-5c3871d89910
go: finding github.com/golang/mock v1.2.0
go: finding github.com/gorilla/rpc v1.2.0+incompatible
go: finding google.golang.org/api v0.3.0
go: finding github.com/hashicorp/go-hclog v0.9.1
go: finding github.com/pmezard/go-difflib v1.0.0
go: finding grpc.go4.org v0.0.0-20170609214715-11d0a25b4919
go: finding github.com/armon/go-metrics v0.0.0-20190430140413-ec5e00d3c878
go: finding github.com/coreos/go-systemd v0.0.0-20181012123002-c6f51f82210d
go: finding github.com/minio/sha256-simd v0.1.0
go: finding github.com/minio/hdfs/v3 v3.0.0
go: finding github.com/coreos/go-systemd v0.0.0-20190321100706-95778dfbb74e
go: finding golang.org/x/net v0.0.0-20190311183353-d8887717615a
go: finding golang.org/x/sys v0.0.0-20181029174526-d69651ed3497
go: finding golang.org/x/net v0.0.0-20181029044818-c44066c5c816
go: finding honnef.co/go/tools v0.0.0-20190106161140-3f1c8253044a
go: finding github.com/mitchellh/cli v1.0.0
go: finding github.com/stretchr/testify v1.3.0
go: finding github.com/gernest/wow v0.1.0
go: finding github.com/klauspost/compress v1.5.0
go: finding github.com/apache/thrift v0.12.0
go: finding google.golang.org/genproto v0.0.0-20180831171423-11092d34479b
go: finding github.com/grpc-ecosystem/grpc-gateway v1.5.0
go: finding google.golang.org/api v0.0.0-20180910000450-7ca32eb868bf
go: finding github.com/googleapis/gax-go v2.0.0+incompatible
go: finding github.com/posener/complete v1.1.1
go: finding github.com/circonus-labs/circonusllhist v0.1.3
go: finding github.com/json-iterator/go v1.1.6
go: finding golang.org/x/sys v0.0.0-20190116161447-11f53e031339
go: finding google.golang.org/appengine v1.2.0
go: finding gopkg.in/jcmturner/aescts.v1 v1.0.1
go: finding github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: finding github.com/nats-io/nkeys v0.0.2
go: finding github.com/hashicorp/go-msgpack v0.5.5
go: finding github.com/googleapis/gax-go/v2 v2.0.4
go: finding github.com/jmespath/go-jmespath v0.0.0-20180206201540-c2b33e8439af
go: finding golang.org/x/crypto v0.0.0-20181203042331-505ab145d0a9
go: finding github.com/prometheus/client_model v0.0.0-20190115171406-56726106282f
go: finding golang.org/x/net v0.0.0-20190108225652-1e06a53dbb7e
go: finding golang.org/x/lint v0.0.0-20190227174305-5b3e6a55c961
go: finding golang.org/x/time v0.0.0-20180412165947-fbb02b2291d2
go: finding github.com/prometheus/procfs v0.0.0-20180725123919-05ee40e3a273
go: finding google.golang.org/genproto v0.0.0-20180817151627-c66870c02cf8
go: finding github.com/pborman/getopt v0.0.0-20180729010549-6fdd0a2c7117
go: finding github.com/stretchr/objx v0.1.0
go: finding github.com/jcmturner/gofork v0.0.0-20180107083740-2aebee971930
go: finding github.com/DataDog/datadog-go v2.2.0+incompatible
go: finding github.com/prometheus/procfs v0.0.0-20190117184657-bf6a532e95b1
go: finding golang.org/x/sys v0.0.0-20190422165155-953cdadca894
go: finding google.golang.org/genproto v0.0.0-20181029155118-b69ba1387ce2
go: finding google.golang.org/grpc v1.14.0
go: finding golang.org/x/crypto v0.0.0-20190103213133-ff983b9c42bc
go: finding github.com/bradfitz/go-smtpd v0.0.0-20170404230938-deb6d6237625
go: finding github.com/hashicorp/go-immutable-radix v1.0.0
go: finding gopkg.in/jcmturner/gokrb5.v5 v5.3.0
go: finding github.com/armon/go-radix v0.0.0-20180808171621-7fddfc383310
go: finding gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: finding github.com/davecgh/go-spew v1.1.1
go: finding github.com/google/go-github v17.0.0+incompatible
go: finding github.com/prometheus/client_golang v0.9.2
go: finding github.com/kr/pretty v0.1.0
go: finding github.com/hashicorp/go-uuid v1.0.0
go: finding golang.org/x/sys v0.0.0-20180823144017-11551d06cbcc
go: finding github.com/pkg/errors v0.8.1
go: finding github.com/tidwall/pretty v0.0.0-20190325153808-1166b9ac2b65
go: finding golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: finding github.com/davecgh/go-spew v1.1.0
go: finding github.com/bgentry/speakeasy v0.1.0
go: finding github.com/openzipkin/zipkin-go v0.1.1
go: finding github.com/gopherjs/gopherjs v0.0.0-20190328170749-bb2674552d8f
go: finding golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: finding github.com/soheilhy/cmux v0.1.4
go: finding github.com/prometheus/client_golang v0.8.0
go: finding github.com/gogo/protobuf v1.2.1
go: finding gopkg.in/cheggaaa/pb.v1 v1.0.28
go: finding github.com/howeyc/gopass v0.0.0-20170109162249-bf9dde6d0d2c
go: finding github.com/klauspost/cpuid v1.2.0
go: finding go.etcd.io/bbolt v1.3.2
go: finding github.com/hashicorp/go-cleanhttp v0.5.1
go: finding golang.org/x/net v0.0.0-20190125091013-d26f9f9a57f3
go: finding gopkg.in/jcmturner/goidentity.v2 v2.0.0
go: finding github.com/tarm/serial v0.0.0-20180830185346-98f6abe2eb07
go: finding github.com/minio/sio v0.0.0-20190118043801-035b4ef8c449
go: finding github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
go: finding gopkg.in/jcmturner/rpc.v0 v0.0.2
go: finding github.com/golang/mock v1.1.1
go: finding github.com/hashicorp/go-multierror v1.0.0
go: finding github.com/hashicorp/go-retryablehttp v0.5.3
go: finding golang.org/x/net v0.0.0-20181201002055-351d144fa1fc
go: finding golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4
go: finding github.com/nats-io/go-nats-streaming v0.4.2
go: finding github.com/gregjones/httpcache v0.0.0-20180305231024-9cad4c3443a7
go: finding github.com/minio/dsync v1.0.0
go: finding go.uber.org/zap v1.9.1
go: finding golang.org/x/oauth2 v0.0.0-20181017192945-9dcd33a902f4
go: finding git.apache.org/thrift.git v0.0.0-20180902110319-2566ecd5d999
go: finding github.com/coreos/bbolt v1.3.2
go: finding github.com/klauspost/compress v1.4.1
go: finding github.com/tmc/grpc-websocket-proxy v0.0.0-20190109142713-0ad062ec5ee5
go: finding google.golang.org/api v0.2.0
go: finding go.uber.org/multierr v1.1.0
go: finding github.com/coreos/pkg v0.0.0-20180928190104-399ea9e2e55f
go: finding go4.org v0.0.0-20180809161055-417644f6feb5
go: finding github.com/golang/groupcache v0.0.0-20190129154638-5b532d6fd5ef
go: finding github.com/segmentio/go-prompt v1.2.1-0.20161017233205-f0d19b6901ad
go: finding github.com/pascaldekloe/goe v0.1.0
go: finding gopkg.in/inf.v0 v0.9.1
go: finding github.com/sirupsen/logrus v1.3.0
go: finding contrib.go.opencensus.io/exporter/ocagent v0.4.10
go: finding github.com/prometheus/procfs v0.0.0-20181204211112-1dc9a6cbc91a
go: finding github.com/mattn/go-runewidth v0.0.4
go: finding github.com/circonus-labs/circonus-gometrics v2.3.1+incompatible
go: finding golang.org/x/net v0.0.0-20180724234803-3673e40ba225
go: finding cloud.google.com/go v0.34.0
go: finding github.com/minio/mc v0.0.0-20190401030144-a1355e50e2e8
go: finding github.com/kr/text v0.1.0
go: finding github.com/minio/blazer v0.0.0-20171126203752-2081f5bf0465
go: finding github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: finding github.com/tv42/httpunix v0.0.0-20150427012821-b75d8614f926
go: finding github.com/hashicorp/go-msgpack v0.5.4
go: finding github.com/jonboulle/clockwork v0.1.0
go: finding github.com/mattn/go-isatty v0.0.3
go: finding golang.org/x/crypto v0.0.0-20190426145343-a29dc8fdc734
go: finding github.com/gorilla/websocket v1.4.0
go: finding golang.org/x/sys v0.0.0-20190130150945-aca44879d564
go: finding github.com/hashicorp/go-uuid v0.0.0-20180228145832-27454136f036
go: finding golang.org/x/sys v0.0.0-20190429190828-d89cdac9e872
go: finding github.com/fortytw2/leaktest v1.2.0
go: finding google.golang.org/grpc v1.19.1
go: finding golang.org/x/perf v0.0.0-20180704124530-6e6d33e29852
go: finding github.com/hashicorp/raft v1.0.1
go: finding github.com/konsorten/go-windows-terminal-sequences v1.0.1
go: finding github.com/kr/pty v1.1.1
go: finding github.com/mitchellh/go-wordwrap v1.0.0
go: finding golang.org/x/sync v0.0.0-20181108010431-42b317875d0f
go: finding gopkg.in/yaml.v2 v2.2.1
go: finding github.com/nats-io/nats v1.7.2
go: finding github.com/anmitsu/go-shlex v0.0.0-20161002113705-648efa622239
go: finding github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2
go: finding github.com/DataDog/zstd v1.3.5
go: finding github.com/minio/lsync v0.0.0-20190207022115-a4e43e3d0887
go: finding github.com/kisielk/errcheck v1.1.0
go: finding golang.org/x/sys v0.0.0-20180905080454-ebe1bf3edb33
go: finding github.com/stretchr/objx v0.1.1
go: finding go.opencensus.io v0.19.1
go: finding github.com/hashicorp/go-cleanhttp v0.5.0
go: finding github.com/ugorji/go/codec v0.0.0-20190320090025-2dc34c0b8780
go: finding github.com/gliderlabs/ssh v0.1.1
go: finding golang.org/x/crypto v0.0.0-20190325154230-a5d413f7728c
go: finding golang.org/x/net v0.0.0-20190424112056-4829fb13d2c6
go: finding github.com/smartystreets/assertions v0.0.0-20190401211740-f487f9de1cd3
go: finding golang.org/x/tools v0.0.0-20190226205152-f727befe758c
go: finding github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: finding golang.org/x/crypto v0.0.0-20180904163835-0709b304e793
go: finding github.com/prometheus/common v0.0.0-20181126121408-4724e9255275
go: finding golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2
go: finding github.com/minio/minio v0.0.0-20190325204105-0250f7de678b
go: finding github.com/keybase/go-crypto v0.0.0-20190312101036-b475f2ecc1fe
go: finding github.com/Nvveen/Gotty v0.0.0-20120604004816-cd527374f1e5
go: finding contrib.go.opencensus.io/exporter/stackdriver v0.0.0-20180919222851-d1e19f5c23e9
go: finding gopkg.in/cheggaaa/pb.v2 v2.0.6
go: finding github.com/go-stomp/stomp v2.0.2+incompatible
go: finding github.com/nats-io/nuid v1.0.0
go: finding github.com/skyrings/skyring-common v0.0.0-20160324141443-762fd2bfc12e
go: finding golang.org/x/crypto v0.0.0-20180723164146-c126467f60eb
go: finding github.com/flynn/go-shlex v0.0.0-20150515145356-3f9db97f8568
go: finding golang.org/x/net v0.0.0-20190324223953-e3b2ff56ed87
go: finding github.com/minio/dsync v0.0.0-20190104003057-61c41ffdeea2
go: finding github.com/gorilla/rpc v0.0.0-20160517062331-bd3317b8f670
go: finding github.com/nats-io/go-nats v1.7.2
go: finding github.com/hashicorp/go-rootcerts v1.0.0
go: finding github.com/klauspost/cpuid v0.0.0-20160106104451-349c67577817
go: finding github.com/hashicorp/raft v1.0.0
go: finding golang.org/x/sys v0.0.0-20190318195719-6c81ef8f67ca
go: finding github.com/elazarl/go-bindata-assetfs v0.0.0-20151224045452-57eb5e1fc594
go: finding google.golang.org/appengine v1.0.0
go: finding github.com/aliyun/alibaba-cloud-sdk-go v0.0.0-20190315122603-6f9e54af456e
go: finding github.com/dimchansky/utfbom v1.1.0
go: finding github.com/minio/cli v0.0.0-20170227073228-b8ae5507c0ce
go: finding gopkg.in/mattn/go-isatty.v0 v0.0.4
go: finding honnef.co/go/tools v0.0.0-20180920025451-e3ad64cb4ed3
go: finding github.com/gotestyourself/gotestyourself v2.2.0+incompatible
go: finding github.com/hashicorp/vault-plugin-secrets-kv v0.0.0-20190315192709-dccffee64925
go: finding golang.org/x/tools v0.0.0-20181219222714-6e267b5cc78e
go: finding google.golang.org/genproto v0.0.0-20181219182458-5a97ab628bfb
go: finding github.com/NYTimes/gziphandler v1.1.1
go: finding github.com/djherbis/atime v0.0.0-20170215084934-89517e96e10b
go: finding github.com/hashicorp/go-version v0.0.0-20160214002439-2e7f5ea8e27b
go: finding github.com/nats-io/nats v0.0.0-20160916181735-70b70be17b77
go: finding github.com/centrify/cloud-golang-sdk v0.0.0-20190214225812-119110094d0f
go: finding github.com/golang/protobuf v1.1.0
go: finding golang.org/x/net v0.0.0-20181106065722-10aee1819953
go: finding github.com/nats-io/go-nats-streaming v0.0.0-20161216191029-077898146bfb
go: finding github.com/tidwall/gjson v1.1.4
go: finding github.com/denisenkom/go-mssqldb v0.0.0-20190315220205-a8ed825ac853
go: finding github.com/opencontainers/image-spec v1.0.1
go: finding k8s.io/klog v0.2.0
go: finding github.com/containerd/continuity v0.0.0-20181203112020-004b46473808
go: finding github.com/lib/pq v0.0.0-20181016162627-9eb73efc1fcc
go: finding github.com/hashicorp/go-hclog v0.8.0
go: finding gopkg.in/fatih/color.v1 v1.7.0
go: finding github.com/klauspost/crc32 v0.0.0-20161016154125-cb6bfca970f6
go: finding github.com/gammazero/deque v0.0.0-20190130191400-2afb3858e9c7
go: finding github.com/cheggaaa/pb v0.0.0-20160713104425-73ae1d68fe0b
go: finding github.com/pquerna/cachecontrol v0.0.0-20180517163645-1555304b9b35
go: finding github.com/minio/highwayhash v0.0.0-20181220011308-93ed73d64169
go: finding github.com/coreos/etcd v0.0.0-20180703215944-e4425ee79f2f
go: finding github.com/hashicorp/go-rootcerts v0.0.0-20160503143440-6bb64b370b90
go: finding github.com/mitchellh/mapstructure v1.1.2
go: finding github.com/hashicorp/vault-plugin-auth-azure v0.0.0-20190201222632-0af1d040b5b3
go: finding github.com/gocql/gocql v0.0.0-20190301043612-f6df8288f9b4
go: finding github.com/minio/mc v0.0.0-20190311071728-2e612b23d665
go: finding github.com/dgrijalva/jwt-go v0.0.0-20180921172315-3af4c746e1c2
go: finding github.com/boombuler/barcode v1.0.0
go: finding github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: finding github.com/hashicorp/vault-plugin-secrets-ad v0.0.0-20190131222416-4796d9980125
go: finding github.com/ory/dockertest v3.3.4+incompatible
go: finding github.com/miekg/dns v1.0.14
go: finding github.com/openzipkin/zipkin-go v0.1.3
go: finding contrib.go.opencensus.io/exporter/ocagent v0.4.7
go: finding github.com/garyburd/redigo v1.6.0
go: finding github.com/go-ini/ini v1.42.0
go: finding github.com/bitly/go-hostpool v0.0.0-20171023180738-a3a6125de932
go: finding github.com/Azure/go-autorest v11.5.2+incompatible
go: finding github.com/hashicorp/vault-plugin-secrets-gcp v0.0.0-20190311200649-621231cb86fe
go: finding github.com/gomodule/redigo v0.0.0-20190205135352-43fe51054af5
go: finding gopkg.in/ory-am/dockertest.v2 v2.2.3
go: finding github.com/pkg/profile v1.2.1
go: finding github.com/smartystreets/goconvey v0.0.0-20190306220146-200a235640ff
go: finding github.com/hashicorp/vault-plugin-auth-jwt v0.0.0-20190314211503-86b44673ce1e
go: finding github.com/gorilla/mux v0.0.0-20160605233521-9fa818a44c2b
go: finding github.com/mattn/go-isatty v0.0.4
go: finding github.com/minio/sio v0.0.0-20180327104954-6a41828a60f0
go: finding gopkg.in/h2non/filetype.v1 v1.0.3
go: finding github.com/mitchellh/copystructure v1.0.0
go: finding github.com/Shopify/toxiproxy v2.1.4+incompatible
go: finding github.com/grpc-ecosystem/grpc-gateway v1.6.2
go: finding github.com/hashicorp/serf v0.8.2
go: finding github.com/eapache/queue v1.1.0
go: finding gopkg.in/VividCortex/ewma.v1 v1.1.1
go: finding golang.org/x/crypto v0.0.0-20190131182504-b8fe1690c613
go: finding github.com/mitchellh/go-testing-interface v1.0.0
go: finding github.com/go-errors/errors v1.0.1
go: finding github.com/mitchellh/hashstructure v1.0.0
go: finding google.golang.org/grpc v1.15.0
go: finding github.com/patrickmn/go-cache v2.1.0+incompatible
go: finding github.com/SermoDigital/jose v0.9.1
go: finding google.golang.org/api v0.0.0-20181220000619-583d854617af
go: finding github.com/ory-am/common v0.4.0
go: finding google.golang.org/api v0.0.0-20180603000442-8e296ef26005
go: finding github.com/mattbaird/elastigo v0.0.0-20170123220020-2fe47fd29e4b
go: finding github.com/golang/protobuf v1.3.0
go: finding github.com/asaskevich/govalidator v0.0.0-20180720115003-f9ffefc3facf
go: finding gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce
go: finding github.com/Microsoft/go-winio v0.4.12
go: finding golang.org/x/oauth2 v0.0.0-20180603041954-1e0a3fa8ba9a
go: finding google.golang.org/appengine v1.3.0
go: finding github.com/spf13/pflag v1.0.3
go: finding github.com/nsqio/go-nsq v0.0.0-20181028195256-0527e80f3ba5
go: finding github.com/tidwall/match v1.0.0
go: finding github.com/hashicorp/go-syslog v1.0.0
go: finding google.golang.org/api v0.0.0-20180916000451-19ff8768a5c0
go: finding github.com/eapache/go-xerial-snappy v0.0.0-20160609142408-bb955e01b934
go: finding golang.org/x/lint v0.0.0-20180702182130-06c8688daad7
go: finding github.com/gorilla/context v0.0.0-20160525203319-aed02d124ae4
go: finding github.com/golang/lint v0.0.0-20180702182130-06c8688daad7
go: finding github.com/coreos/go-semver v0.0.0-20180108230905-e214231b295a
go: finding github.com/gorilla/handlers v0.0.0-20160410185317-66e6c6f01d8d
go: finding github.com/docker/go-connections v0.4.0
go: finding github.com/hashicorp/logutils v1.0.0
go: finding github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869
go: finding github.com/onsi/gomega v1.5.0
go: finding github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78
go: finding golang.org/x/tools v0.0.0-20180828015842-6cd1fcedba52
go: finding golang.org/x/oauth2 v0.0.0-20181203162652-d668ce993890
go: finding github.com/michaelklishin/rabbit-hole v1.5.0
go: finding github.com/ugorji/go v1.1.2
go: finding honnef.co/go/tools v0.0.0-20180728063816-88497007e858
go: finding github.com/duosecurity/duo_api_golang v0.0.0-20190308151101-6c680f768e74
go: finding gopkg.in/mattn/go-colorable.v0 v0.1.0
go: finding github.com/gogo/protobuf v1.2.0
go: finding github.com/hashicorp/vault-plugin-secrets-azure v0.0.0-20181207232500-0087bdef705a
go: finding github.com/grpc-ecosystem/grpc-gateway v1.7.0
go: finding go.opencensus.io v0.15.0
go: finding github.com/nats-io/nats-streaming-server v0.14.1
go: finding github.com/hashicorp/vault-plugin-secrets-alicloud v0.0.0-20190131211812-b0abe36195cb
go: finding k8s.io/apimachinery v0.0.0-20190313115320-c9defaaddf6f
go: finding github.com/coredns/coredns v0.0.0-20180121192821-d4bf076ccf4e
go: finding github.com/nats-io/nats-streaming-server v0.12.2
go: finding github.com/nats-io/go-nats v0.0.0-20161120202126-6b6bf392d34d
go: finding golang.org/x/sys v0.0.0-20181218192612-074acd46bca6
go: finding github.com/hashicorp/vault-plugin-auth-centrify v0.0.0-20180816201131-66b0a34a58bf
go: finding github.com/minio/sha256-simd v0.0.0-20190131020904-2d45a736cd16
go: finding github.com/fatih/structs v0.0.0-20181010231757-878a968ab225
go: finding google.golang.org/grpc v1.17.0
go: finding github.com/hashicorp/vault-plugin-auth-gcp v0.0.0-20190201215414-7d4c2101e7d0
go: finding github.com/mattn/go-colorable v0.0.0-20160210001857-9fdad7c47650
go: finding github.com/samuel/go-zookeeper v0.0.0-20180130194729-c4fab1ac1bec
go: finding github.com/hashicorp/vault v0.0.0-20181121181053-d4367e581fe1
go: finding github.com/onsi/ginkgo v1.8.0
go: finding github.com/opencontainers/runc v0.1.1
go: finding github.com/hashicorp/go-cleanhttp v0.0.0-20171218145408-d5fe4b57a186
go: finding github.com/prometheus/common v0.2.0
go: finding github.com/hashicorp/go-plugin v0.0.0-20190220160451-3f118e8ee104
go: finding github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
go: finding github.com/aliyun/aliyun-oss-go-sdk v0.0.0-20170925032315-6fe16293d6b7
go: finding github.com/mitchellh/reflectwalk v1.0.0
go: finding github.com/hashicorp/mdns v1.0.0
go: finding github.com/coreos/go-semver v0.2.0
go: finding github.com/jefferai/jsonx v1.0.0
go: finding github.com/mitchellh/go-homedir v1.0.0
go: finding golang.org/x/sys v0.0.0-20190204203706-41f3e6584952
go: finding github.com/mitchellh/pointerstructure v0.0.0-20170205204203-f2329fcfa9e2
go: finding github.com/hashicorp/go-memdb v0.0.0-20190306140544-eea0b16292ad
go: finding gopkg.in/fsnotify.v1 v1.4.7
go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: finding github.com/aws/aws-sdk-go v1.15.31
go: finding github.com/eapache/go-resiliency v0.0.0-20160104191539-b86b1ec0dd42
go: finding gopkg.in/gorethink/gorethink.v4 v4.1.0
go: finding gopkg.in/square/go-jose.v2 v2.3.0
go: finding github.com/hashicorp/go-hclog v0.0.0-20180709165350-ff2cf002a8dd
go: finding github.com/jeffchao/backoff v0.0.0-20140404060208-9d7fd7aa17f2
go: finding layeh.com/radius v0.0.0-20190118135028-0f678f039617
go: finding golang.org/x/sys v0.0.0-20190129075346-302c3dd5f1cc
go: finding github.com/dustin/go-humanize v0.0.0-20150809201405-1c212aae1d02
go: finding github.com/golang/snappy v0.0.0-20170215233205-553a64147049
go: finding google.golang.org/genproto v0.0.0-20180918203901-c3f76f3b92d1
go: finding github.com/cenkalti/backoff v2.1.1+incompatible
go: finding golang.org/x/sys v0.0.0-20181116152217-5ac8a444bdc5
go: finding github.com/google/gofuzz v0.0.0-20170612174753-24818f796faf
go: finding github.com/ryanuber/columnize v0.0.0-20160712163229-9b3edd62028f
go: finding github.com/sirupsen/logrus v1.2.0
go: finding golang.org/x/net v0.0.0-20181023162649-9b4f9f5ad519
go: finding github.com/onsi/ginkgo v1.6.0
go: finding github.com/minio/minio v0.0.0-20190206103305-fd4e15c11641
go: finding google.golang.org/grpc v1.12.0
go: finding github.com/araddon/gou v0.0.0-20190110011759-c797efecbb61
go: finding github.com/pborman/uuid v1.2.0
go: finding sigs.k8s.io/yaml v1.1.0
go: finding github.com/briankassouf/jose v0.9.1
go: finding github.com/fsnotify/fsnotify v1.4.7
go: finding github.com/hashicorp/go-gcp-common v0.0.0-20180425173946-763e39302965
go: finding github.com/coreos/go-systemd v0.0.0-20181031085051-9002847aa142
go: finding github.com/hashicorp/go-msgpack v0.5.3
go: finding github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da
go: finding github.com/mitchellh/iochan v1.0.0
go: finding github.com/Jeffail/gabs v1.1.1
go: finding github.com/cenkalti/backoff v2.0.0+incompatible
go: finding github.com/eclipse/paho.mqtt.golang v0.0.0-20181129145454-379fd9f99ba5
go: finding golang.org/x/net v0.0.0-20180530234432-1e491301e022
go: finding github.com/minio/minio-go v0.0.0-20190313212832-5d20267d970d
go: finding github.com/julienschmidt/httprouter v1.2.0
go: finding github.com/hashicorp/memberlist v0.1.3
go: finding github.com/gopherjs/gopherjs v0.0.0-20190309154008-847fc94819f9
go: finding github.com/Azure/azure-sdk-for-go v26.4.0+incompatible
go: finding golang.org/x/tools v0.0.0-20190318200714-bb1270c20edf
go: finding github.com/gammazero/workerpool v0.0.0-20181230203049-86a96b5d5d92
go: finding github.com/go-test/deep v1.0.1
go: finding golang.org/x/crypto v0.0.0-20180820150726-614d502a4dac
go: finding github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf
go: finding gopkg.in/alecthomas/kingpin.v2 v2.2.6
go: finding golang.org/x/net v0.0.0-20190318221613-d196dffd7c2b
go: finding golang.org/x/net v0.0.0-20181114220301-adae6a3d119a
go: finding github.com/mwitkow/go-conntrack v0.0.0-20161129095857-cc309e4a2223
go: finding github.com/minio/cli v1.3.0
go: finding github.com/hashicorp/yamux v0.0.0-20180604194846-3520598351bb
go: finding github.com/rs/cors v0.0.0-20190116175910-76f58f330d76
go: finding github.com/mitchellh/gox v0.4.0
go: finding github.com/go-ini/ini v1.25.4
go: finding github.com/grpc-ecosystem/go-grpc-middleware v1.0.0
go: finding golang.org/x/sys v0.0.0-20181026203630-95b1ffbd15a5
go: finding gopkg.in/Shopify/sarama.v1 v1.10.1
go: finding github.com/coreos/go-oidc v2.0.0+incompatible
go: finding golang.org/x/lint v0.0.0-20181217174547-8f45f776aaf1
go: finding go.opencensus.io v0.18.1-0.20181204023538-aab39bd6a98b
go: finding github.com/gogo/protobuf v1.1.1
go: finding golang.org/x/tools v0.0.0-20190328211700-ab21143f2384
go: finding github.com/oklog/run v1.0.0
go: finding github.com/sean-/seed v0.0.0-20170313163322-e2103e2c3529
go: finding github.com/segmentio/go-prompt v0.0.0-20161017233205-f0d19b6901ad
go: finding cloud.google.com/go v0.23.0
go: finding github.com/mattn/go-colorable v0.0.9
go: finding github.com/go-ldap/ldap v3.0.2+incompatible
go: finding gopkg.in/mattn/go-runewidth.v0 v0.0.4
go: finding github.com/hashicorp/vault-plugin-auth-alicloud v0.0.0-20190311155555-98628998247d
go: finding github.com/openzipkin/zipkin-go v0.1.6
go: finding github.com/go-kit/kit v0.8.0
go: finding github.com/sirupsen/logrus v1.0.6
go: finding github.com/chrismalek/oktasdk-go v0.0.0-20181212195951-3430665dfaa0
go: finding golang.org/x/tools v0.0.0-20180221164845-07fd8470d635
go: finding github.com/onsi/gomega v1.4.3
go: finding gopkg.in/olivere/elastic.v5 v5.0.31
go: finding github.com/dustin/go-humanize v0.0.0-20170228161531-259d2a102b87
go: finding gopkg.in/fatih/pool.v2 v2.0.0
go: finding github.com/tidwall/gjson v1.1.2
go: finding github.com/klauspost/pgzip v0.0.0-20180606150939-90b2c57fba35
go: finding github.com/kisielk/gotool v1.0.0
go: finding gopkg.in/cheggaaa/pb.v1 v1.0.27
go: finding github.com/hashicorp/nomad v0.8.7
go: finding github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: finding github.com/klauspost/reedsolomon v0.0.0-20190210214925-2b210cf0866d
go: finding golang.org/x/crypto v0.0.0-20181029021203-45a5f77698d3
go: finding github.com/gorilla/mux v1.6.2
go: finding github.com/klauspost/compress v1.3.0
go: finding github.com/hashicorp/go.net v0.0.1
go: finding github.com/hpcloud/tail v1.0.0
go: finding github.com/rcrowley/go-metrics v0.0.0-20181016184325-3113b8401b8a
go: finding github.com/hailocab/go-hostpool v0.0.0-20160125115350-e80d13ce29ed
go: finding github.com/mitchellh/go-testing-interface v0.0.0-20171004221916-a61a99592b77
go: finding github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc
go: finding golang.org/x/sys v0.0.0-20190322080309-f49334f85ddc
go: finding github.com/dancannon/gorethink v4.0.0+incompatible
go: finding github.com/mitchellh/mapstructure v0.0.0-20160808181253-ca63d7c062ee
go: finding github.com/pkg/profile v0.0.0-20160315012113-19396ecfda37
go: finding github.com/opencontainers/go-digest v1.0.0-rc1
go: finding github.com/eapache/go-resiliency v1.1.0
go: finding github.com/kr/logfmt v0.0.0-20140226030751-b84e30acd515
go: finding github.com/hashicorp/go-retryablehttp v0.0.0-20180718195005-e651d75abec6
go: finding github.com/prometheus/client_golang v0.9.1
go: finding github.com/pkg/xattr v0.0.0-20170808190211-56ed87199eba
go: finding github.com/streadway/amqp v0.0.0-20160311215503-2e25825abdbd
go: finding github.com/hashicorp/vault-plugin-secrets-gcpkms v0.0.0-20190116164938-d6b25b0b4a39
go: finding github.com/gorhill/cronexpr v0.0.0-20180427100037-88b0669f7d75
go: finding github.com/onsi/ginkgo v1.7.0
go: finding github.com/SAP/go-hdb v0.14.0
go: finding github.com/coreos/etcd v0.0.0-20180622175804-8f6348a97d48
go: finding github.com/google/uuid v1.0.0
go: finding google.golang.org/grpc v1.16.0
go: finding github.com/golang/snappy v0.0.0-20180518054509-2e65f85255db
go: finding google.golang.org/grpc v1.18.0
go: finding gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: finding google.golang.org/genproto v0.0.0-20180601223552-81158efcc9f2
go: finding github.com/Shopify/sarama v1.19.0
go: finding k8s.io/api v0.0.0-20190313115550-3c12c96769cc
go: finding github.com/pascaldekloe/goe v0.0.0-20180627143212-57f6aae5913c
go: finding github.com/go-stack/stack v1.8.0
go: finding github.com/go-sql-driver/mysql v0.0.0-20181218123637-c45f530f8e7f
go: finding golang.org/x/crypto v0.0.0-20190313024323-a1f597ede03a
go: finding golang.org/x/sys v0.0.0-20180828065106-d99a578cf41b
go: finding github.com/prometheus/procfs v0.0.0-20181005140218-185b4288413d
go: finding github.com/gorilla/context v1.1.1
go: finding github.com/docker/go-units v0.3.3
go: finding github.com/hashicorp/vault-plugin-auth-kubernetes v0.0.0-20190201222209-db96aa4ab438
go: finding gotest.tools v2.2.0+incompatible
go: finding golang.org/x/sys v0.0.0-20181122145206-62eef0e2fa9b
go: finding github.com/fullsailor/pkcs7 v0.0.0-20180613152042-8306686428a5
go: finding github.com/hashicorp/go-sockaddr v1.0.0
go: finding github.com/go-logfmt/logfmt v0.3.0
go: finding github.com/hashicorp/consul v1.4.3
go: finding github.com/opentracing/opentracing-go v1.0.2
go: finding github.com/jmespath/go-jmespath v0.0.0-20160202185014-0b12d6b521d8
go: finding github.com/alecthomas/participle v0.0.0-20190103085315-bf8340a459bd
go: finding github.com/ugorji/go v0.0.0-20180628102755-7d51bbe6161d
go: finding github.com/pquerna/otp v1.1.0
go: downloading github.com/minio/hdfs/v3 v3.0.0
go: downloading github.com/minio/minio-go v0.0.0-20190327203652-5325257a208f
go: downloading github.com/Azure/azure-sdk-for-go v27.0.0+incompatible
go: downloading google.golang.org/api v0.4.0
go: downloading cloud.google.com/go v0.37.2
go: downloading github.com/minio/cli v1.21.0
go: downloading github.com/mattn/go-isatty v0.0.7
go: downloading github.com/golang/snappy v0.0.1
go: extracting github.com/mattn/go-isatty v0.0.7
go: downloading github.com/Azure/go-autorest v11.7.0+incompatible
go: extracting github.com/minio/cli v1.21.0
go: extracting github.com/golang/snappy v0.0.1
go: downloading github.com/fatih/color v1.7.0
go: downloading github.com/minio/lsync v1.0.1
go: extracting github.com/minio/lsync v1.0.1
go: downloading github.com/tidwall/gjson v1.2.1
go: extracting github.com/tidwall/gjson v1.2.1
go: extracting github.com/Azure/go-autorest v11.7.0+incompatible
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: extracting github.com/minio/minio-go v0.0.0-20190327203652-5325257a208f
go: extracting github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/minio/sha256-simd v0.1.0
go: downloading github.com/minio/minio-go/v6 v6.0.29
go: extracting github.com/minio/sha256-simd v0.1.0
go: downloading golang.org/x/crypto v0.0.0-20190621222207-cc06ce4a13d4
go: extracting github.com/fatih/color v1.7.0
go: downloading github.com/nats-io/stan.go v0.4.5
go: extracting github.com/nats-io/stan.go v0.4.5
go: downloading github.com/tidwall/pretty v1.0.0
go: extracting github.com/minio/minio-go/v6 v6.0.29
go: extracting github.com/tidwall/pretty v1.0.0
go: downloading github.com/nats-io/nats.go v1.8.0
go: downloading github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/elazarl/go-bindata-assetfs v1.0.0
go: extracting github.com/elazarl/go-bindata-assetfs v1.0.0
go: downloading github.com/klauspost/readahead v1.3.0
go: extracting github.com/klauspost/readahead v1.3.0
go: downloading github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: extracting github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/minio/highwayhash v1.0.0
go: extracting github.com/minio/hdfs/v3 v3.0.0
go: extracting github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: extracting github.com/minio/highwayhash v1.0.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.1.2-0.20190322152051-20337d8c3947
go: downloading github.com/go-sql-driver/mysql v1.4.1
go: extracting github.com/nats-io/nats.go v1.8.0
go: downloading gopkg.in/olivere/elastic.v5 v5.0.80
go: downloading github.com/coreos/etcd v3.3.12+incompatible
go: extracting github.com/eclipse/paho.mqtt.golang v1.1.2-0.20190322152051-20337d8c3947
go: extracting github.com/go-sql-driver/mysql v1.4.1
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/minio/mc v0.0.0-20190529152718-f4bb0b8850cb
go: extracting github.com/ncw/directio v1.0.5
go: downloading gopkg.in/yaml.v2 v2.2.2
go: extracting github.com/minio/mc v0.0.0-20190529152718-f4bb0b8850cb
go: downloading golang.org/x/sys v0.0.0-20190626221950-04f50cda93cb
go: extracting golang.org/x/crypto v0.0.0-20190621222207-cc06ce4a13d4
go: downloading github.com/cheggaaa/pb v1.0.28
go: extracting github.com/cheggaaa/pb v1.0.28
go: downloading gopkg.in/jcmturner/gokrb5.v5 v5.3.0
go: extracting gopkg.in/yaml.v2 v2.2.2
go: downloading github.com/json-iterator/go v1.1.6
go: extracting github.com/json-iterator/go v1.1.6
go: downloading github.com/gorilla/rpc v1.2.0+incompatible
go: extracting github.com/gorilla/rpc v1.2.0+incompatible
go: downloading github.com/mattn/go-colorable v0.1.1
go: extracting github.com/mattn/go-colorable v0.1.1
go: downloading github.com/klauspost/reedsolomon v1.9.1
go: extracting gopkg.in/olivere/elastic.v5 v5.0.80
go: extracting github.com/klauspost/reedsolomon v1.9.1
go: downloading github.com/djherbis/atime v1.0.0
go: extracting github.com/djherbis/atime v1.0.0
go: downloading github.com/tidwall/match v1.0.1
go: extracting github.com/tidwall/match v1.0.1
go: downloading github.com/gogo/protobuf v1.2.1
go: downloading github.com/mattn/go-runewidth v0.0.4
go: extracting github.com/mattn/go-runewidth v0.0.4
go: downloading contrib.go.opencensus.io/exporter/ocagent v0.5.0
go: extracting cloud.google.com/go v0.37.2
go: extracting golang.org/x/sys v0.0.0-20190626221950-04f50cda93cb
go: downloading gopkg.in/Shopify/sarama.v1 v1.20.0
go: extracting contrib.go.opencensus.io/exporter/ocagent v0.5.0
go: downloading github.com/streadway/amqp v0.0.0-20190402114354-16ed540749f6
go: downloading go.opencensus.io v0.21.0
go: extracting github.com/streadway/amqp v0.0.0-20190402114354-16ed540749f6
go: extracting github.com/coreos/etcd v3.3.12+incompatible
go: downloading github.com/klauspost/cpuid v1.2.1
go: extracting github.com/klauspost/cpuid v1.2.1
go: downloading github.com/rs/cors v1.6.0
go: extracting github.com/rs/cors v1.6.0
go: downloading github.com/golang/protobuf v1.3.1
go: downloading github.com/pkg/profile v1.3.0
go: extracting github.com/pkg/profile v1.3.0
go: extracting go.opencensus.io v0.21.0
go: extracting gopkg.in/Shopify/sarama.v1 v1.20.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: extracting github.com/dustin/go-humanize v1.0.0
go: downloading github.com/modern-go/reflect2 v1.0.1
go: extracting github.com/modern-go/reflect2 v1.0.1
go: downloading github.com/census-instrumentation/opencensus-proto v0.2.0
go: extracting github.com/golang/protobuf v1.3.1
go: extracting github.com/census-instrumentation/opencensus-proto v0.2.0
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: extracting github.com/gomodule/redigo v2.0.0+incompatible
go: downloading google.golang.org/grpc v1.20.1
go: extracting github.com/davecgh/go-spew v1.1.1
go: extracting github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: downloading github.com/aliyun/aliyun-oss-go-sdk v0.0.0-20190307165228-86c17b95fcd5
go: extracting github.com/aliyun/aliyun-oss-go-sdk v0.0.0-20190307165228-86c17b95fcd5
go: extracting google.golang.org/grpc v1.20.1
go: extracting github.com/gogo/protobuf v1.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/minio/dsync/v2 v2.0.0
go: extracting github.com/mitchellh/go-homedir v1.1.0
go: downloading golang.org/x/net v0.0.0-20190620200207-3b0461eec859
go: extracting github.com/minio/dsync/v2 v2.0.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: extracting github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/nats-io/nkeys v0.0.2
go: extracting github.com/nats-io/nkeys v0.0.2
go: downloading github.com/nats-io/nuid v1.0.1
go: extracting github.com/nats-io/nuid v1.0.1
go: downloading github.com/gorilla/handlers v1.4.0
go: extracting github.com/gorilla/handlers v1.4.0
go: downloading github.com/googleapis/gax-go v2.0.0+incompatible
go: extracting github.com/googleapis/gax-go v2.0.0+incompatible
go: downloading github.com/googleapis/gax-go/v2 v2.0.4
go: extracting github.com/googleapis/gax-go/v2 v2.0.4
go: extracting golang.org/x/net v0.0.0-20190620200207-3b0461eec859
go: downloading github.com/klauspost/pgzip v1.2.1
go: extracting github.com/klauspost/pgzip v1.2.1
go: downloading github.com/gorilla/mux v1.7.0
go: extracting github.com/gorilla/mux v1.7.0
go: downloading google.golang.org/genproto v0.0.0-20190307195333-5fe7a883aa19
go: finding github.com/eapache/go-xerial-snappy latest
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: extracting github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/hashicorp/golang-lru v0.5.1
go: extracting github.com/hashicorp/golang-lru v0.5.1
go: downloading golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: extracting golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: downloading github.com/klauspost/compress v1.5.0
go: extracting google.golang.org/genproto v0.0.0-20190307195333-5fe7a883aa19
go: extracting github.com/klauspost/compress v1.5.0
go: downloading gopkg.in/ini.v1 v1.42.0
go: extracting gopkg.in/ini.v1 v1.42.0
go: downloading github.com/lib/pq v1.0.0
go: extracting google.golang.org/api v0.4.0
go: extracting github.com/lib/pq v1.0.0
go: downloading github.com/pkg/errors v0.8.1
go: extracting github.com/pkg/errors v0.8.1
go: downloading github.com/minio/parquet-go v0.0.0-20190318185229-9d767baf1679
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: extracting github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading github.com/rjeczalik/notify v0.9.2
go: extracting github.com/alecthomas/participle v0.2.1
go: extracting github.com/minio/parquet-go v0.0.0-20190318185229-9d767baf1679
go: downloading github.com/coredns/coredns v1.4.0
go: downloading golang.org/x/text v0.3.2
go: downloading github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
go: extracting github.com/rjeczalik/notify v0.9.2
go: downloading github.com/skyrings/skyring-common v0.0.0-20160929130248-d1c0bb1cbd5e
go: extracting github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
go: downloading go.uber.org/atomic v1.3.2
go: downloading github.com/eapache/queue v1.1.0
go: extracting github.com/eapache/queue v1.1.0
go: downloading github.com/miekg/dns v1.1.8
go: extracting go.uber.org/atomic v1.3.2
go: extracting github.com/miekg/dns v1.1.8
go: downloading git.apache.org/thrift.git v0.12.0
go: extracting github.com/skyrings/skyring-common v0.0.0-20160929130248-d1c0bb1cbd5e
go: downloading golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
go: extracting golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
go: downloading github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: extracting github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: downloading golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: extracting golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: extracting github.com/coredns/coredns v1.4.0
go: downloading github.com/prometheus/common v0.2.0
go: extracting github.com/prometheus/common v0.2.0
go: downloading github.com/fatih/structs v1.1.0
go: extracting github.com/fatih/structs v1.1.0
go: downloading github.com/eapache/go-resiliency v1.2.0
go: extracting github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: extracting github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: downloading github.com/grpc-ecosystem/grpc-gateway v1.9.0
go: extracting gopkg.in/jcmturner/gokrb5.v5 v5.3.0
go: extracting github.com/grpc-ecosystem/grpc-gateway v1.9.0
go: downloading github.com/prometheus/procfs v0.0.0-20190117184657-bf6a532e95b1
go: extracting github.com/prometheus/procfs v0.0.0-20190117184657-bf6a532e95b1
go: downloading github.com/hashicorp/vault v1.1.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: downloading github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: downloading github.com/pierrec/lz4 v2.0.5+incompatible
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: extracting github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: extracting github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/minio/sio v0.2.0
go: downloading github.com/tidwall/sjson v1.0.4
go: extracting github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: downloading github.com/hashicorp/go-uuid v1.0.1
go: extracting github.com/tidwall/sjson v1.0.4
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: extracting github.com/hashicorp/go-uuid v1.0.1
go: downloading github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: extracting github.com/minio/sio v0.2.0
go: downloading github.com/kurin/blazer v0.5.4-0.20190613185654-cf2f27cc0be3
go: extracting github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: extracting github.com/kurin/blazer v0.5.4-0.20190613185654-cf2f27cc0be3
go: downloading gopkg.in/jcmturner/rpc.v0 v0.0.2
go: extracting gopkg.in/jcmturner/aescts.v1 v1.0.1
go: extracting gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: extracting gopkg.in/jcmturner/rpc.v0 v0.0.2
go: extracting git.apache.org/thrift.git v0.12.0
go: extracting github.com/Azure/azure-sdk-for-go v27.0.0+incompatible
go: extracting github.com/pierrec/lz4 v2.0.5+incompatible
go: extracting golang.org/x/text v0.3.2
go: downloading github.com/marstr/guid v1.1.0
go: downloading github.com/satori/go.uuid v1.2.0
go: extracting github.com/marstr/guid v1.1.0
go: extracting github.com/satori/go.uuid v1.2.0
go: extracting github.com/hashicorp/vault v1.1.0
go: downloading github.com/hashicorp/errwrap v1.0.0
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/hashicorp/go-cleanhttp v0.5.1
go: downloading github.com/hashicorp/go-rootcerts v1.0.1
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading github.com/hashicorp/go-retryablehttp v0.5.3
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: extracting github.com/hashicorp/go-cleanhttp v0.5.1
go: extracting github.com/ryanuber/go-glob v1.0.0
go: extracting github.com/hashicorp/errwrap v1.0.0
go: extracting github.com/hashicorp/go-retryablehttp v0.5.3
go: extracting github.com/hashicorp/go-multierror v1.0.0
go: extracting github.com/hashicorp/go-rootcerts v1.0.1
go: extracting github.com/hashicorp/hcl v1.0.0
go: downloading github.com/mitchellh/mapstructure v1.1.2
go: extracting github.com/mitchellh/mapstructure v1.1.2
go: extracting github.com/hashicorp/go-sockaddr v1.0.2
# github.com/minio/minio/cmd
cmd/iam.go:852:25: sys.policyDBSet(objectAPI, accessKey, uinfo.PolicyName) used as value
cmd/iam.go:909:24: sys.policyDBSet(objectAPI, name, policy) used as value
cmd/iam.go:916:3: too many arguments to return
	have (error)
	want ()
cmd/iam.go:920:3: too many arguments to return
	have (error)
	want ()
cmd/iam.go:925:3: too many arguments to return
	have (error)
	want ()
cmd/iam.go:931:3: too many arguments to return
	have (error)
	want ()
cmd/iam.go:941:3: too many arguments to return
	have (error)
	want ()
cmd/iam.go:944:2: too many arguments to return
	have (nil)
	want ()
make: *** [build] Error 2

```

--- a/cmd/admin-handlers.go
+++ b/cmd/admin-handlers.go
@@ -1253,7 +1253,7 @@ func (a adminAPIHandlers) SetUserPolicy(w http.ResponseWriter, r *http.Request) 
 		return
 	}
 
-	if err := globalIAMSys.SetUserPolicy(accessKey, policyName); err != nil {
+	if err := globalIAMSys.PolicyDBSet(accessKey, policyName); err != nil {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 	}
 

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -66,7 +66,7 @@ const (
 )
 
 type iamFormat struct {
-	Version int `json:"format"`
+	Version int `json:"version"`
 }
 
 func getCurrentIAMFormat() iamFormat {

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -20,6 +20,7 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"errors"
 	"path"
 	"strings"
 	"sync"
@@ -47,19 +48,279 @@ const (
 	// IAM sts directory.
 	iamConfigSTSPrefix = iamConfigPrefix + "/sts/"
 
+	// IAM Policy DB prefixes.
+	iamConfigPolicyDBPrefix         = iamConfigPrefix + "/policydb/"
+	iamConfigPolicyDBUsersPrefix    = iamConfigPolicyDBPrefix + "users/"
+	iamConfigPolicyDBSTSUsersPrefix = iamConfigPolicyDBPrefix + "sts-users/"
+
 	// IAM identity file which captures identity credentials.
 	iamIdentityFile = "identity.json"
 
 	// IAM policy file which provides policies for each users.
 	iamPolicyFile = "policy.json"
+
+	// IAM format file
+	iamFormatFile = "format.json"
+
+	iamFormatVersion1 = 1
 )
+
+type iamFormat struct {
+	Version int `json:"format"`
+}
+
+func getCurrentIAMFormat() iamFormat {
+	return iamFormat{Version: iamFormatVersion1}
+}
+
+func getIAMFormatFilePath() string {
+	return iamConfigPrefix + "/" + iamFormatFile
+}
+
+func getUserIdentityPath(user string, isSTS bool) string {
+	basePath := iamConfigUsersPrefix
+	if isSTS {
+		basePath = iamConfigSTSPrefix
+	}
+	return pathJoin(basePath, user, iamIdentityFile)
+}
+
+func getPolicyDocPath(name string) string {
+	return pathJoin(iamConfigPoliciesPrefix, name, iamPolicyFile)
+}
+
+func getMappedPolicyPath(name string, isSTS bool) string {
+	if isSTS {
+		return pathJoin(iamConfigPolicyDBSTSUsersPrefix, name+".json")
+	}
+	return pathJoin(iamConfigPolicyDBUsersPrefix, name+".json")
+}
+
+// MappedPolicy represents a policy name mapped to a user or group
+type MappedPolicy struct {
+	Version int    `json:"version"`
+	Policy  string `json:"policy"`
+}
+
+func newMappedPolicy(policy string) MappedPolicy {
+	return MappedPolicy{Version: 1, Policy: policy}
+}
+
+// UserIdentity represents a user's secret key and their status
+type UserIdentity struct {
+	Version     int              `json:"version"`
+	Credentials auth.Credentials `json:"credentials"`
+}
+
+func newUserIdentity(creds auth.Credentials) UserIdentity {
+	return UserIdentity{Version: 1, Credentials: creds}
+}
+
+func loadIAMConfigItem(objectAPI ObjectLayer, item interface{}, path string) error {
+	data, err := readConfig(context.Background(), objectAPI, path)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(data, item)
+}
+
+func saveIAMConfigItem(objectAPI ObjectLayer, item interface{}, path string) error {
+	data, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	return saveConfig(context.Background(), objectAPI, path, data)
+}
+
+// helper type for listIAMConfigItems
+type itemOrErr struct {
+	Item string
+	Err  error
+}
+
+// Lists files or dirs in the minioMetaBucket at the given path
+// prefix. If dirs is true, only directories are listed, otherwise
+// only objects are listed. All returned items have the pathPrefix
+// removed from their names.
+func listIAMConfigItems(objectAPI ObjectLayer, pathPrefix string, dirs bool,
+	doneCh <-chan struct{}) <-chan itemOrErr {
+
+	ch := make(chan itemOrErr)
+	dirList := func(lo ListObjectsInfo) []string {
+		return lo.Prefixes
+	}
+	filesList := func(lo ListObjectsInfo) (r []string) {
+		for _, o := range lo.Objects {
+			r = append(r, o.Name)
+		}
+		return r
+	}
+
+	go func() {
+		marker := ""
+		for {
+			lo, err := objectAPI.ListObjects(context.Background(),
+				minioMetaBucket, pathPrefix, marker, "/", 1000)
+			if err != nil {
+				select {
+				case ch <- itemOrErr{Err: err}:
+				case <-doneCh:
+				}
+				close(ch)
+				return
+			}
+			marker = lo.NextMarker
+			lister := dirList(lo)
+			if !dirs {
+				lister = filesList(lo)
+			}
+			for _, itemPrefix := range lister {
+				item := strings.TrimPrefix(itemPrefix, pathPrefix)
+				item = strings.TrimSuffix(item, "/")
+				select {
+				case ch <- itemOrErr{Item: item}:
+				case <-doneCh:
+					close(ch)
+					return
+				}
+			}
+			if !lo.IsTruncated {
+				close(ch)
+				return
+			}
+		}
+	}()
+	return ch
+}
+
+func loadIAMConfigItemEtcd(ctx context.Context, item interface{}, path string) error {
+	pdata, err := readConfigEtcd(ctx, globalEtcdClient, path)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(pdata, item)
+}
+
+func saveIAMConfigItemEtcd(ctx context.Context, item interface{}, path string) error {
+	data, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	return saveConfigEtcd(context.Background(), globalEtcdClient, path, data)
+}
 
 // IAMSys - config system.
 type IAMSys struct {
 	sync.RWMutex
-	iamUsersMap        map[string]auth.Credentials
-	iamPolicyMap       map[string]string
-	iamCannedPolicyMap map[string]iampolicy.Policy
+	// map of usernames to credentials
+	iamUsersMap map[string]auth.Credentials
+	// map of policy names to policy definitions
+	iamPolicyDocsMap map[string]iampolicy.Policy
+	// map of usernames/temporary access keys to policy names
+	iamUserPolicyMap map[string]MappedPolicy
+}
+
+func loadPolicyDoc(objectAPI ObjectLayer, policy string, m map[string]iampolicy.Policy) error {
+	var p iampolicy.Policy
+	err := loadIAMConfigItem(objectAPI, &p, getPolicyDocPath(policy))
+	if err != nil {
+		return err
+	}
+	m[policy] = p
+	return nil
+}
+
+func loadPolicyDocs(objectAPI ObjectLayer, m map[string]iampolicy.Policy) error {
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+	for item := range listIAMConfigItems(objectAPI, iamConfigPoliciesPrefix, true, doneCh) {
+		if item.Err != nil {
+			return item.Err
+		}
+
+		policyName := item.Item
+		err := loadPolicyDoc(objectAPI, policyName, m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func loadUser(objectAPI ObjectLayer, user string, isSTS bool,
+	m map[string]auth.Credentials) error {
+
+	var u UserIdentity
+	err := loadIAMConfigItem(objectAPI, &u, getUserIdentityPath(user, isSTS))
+	if err != nil {
+		return err
+	}
+
+	if u.Credentials.IsExpired() {
+		idFile := getUserIdentityPath(user, isSTS)
+		// Delete expired identity - ignoring errors here.
+		deleteConfig(context.Background(), objectAPI, idFile)
+		deleteConfig(context.Background(), objectAPI, getMappedPolicyPath(user, isSTS))
+		return nil
+	}
+
+	m[user] = u.Credentials
+	return nil
+}
+
+func loadUsers(objectAPI ObjectLayer, isSTS bool, m map[string]auth.Credentials) error {
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+	basePrefix := iamConfigUsersPrefix
+	if isSTS {
+		basePrefix = iamConfigSTSPrefix
+	}
+	for item := range listIAMConfigItems(objectAPI, basePrefix, true, doneCh) {
+		if item.Err != nil {
+			return item.Err
+		}
+
+		userName := item.Item
+		err := loadUser(objectAPI, userName, isSTS, m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func loadMappedPolicy(objectAPI ObjectLayer, name string, isSTS bool,
+	m map[string]MappedPolicy) error {
+
+	var p MappedPolicy
+	err := loadIAMConfigItem(objectAPI, &p, getMappedPolicyPath(name, isSTS))
+	if err != nil {
+		return err
+	}
+	m[name] = p
+	return nil
+}
+
+func loadMappedPolicies(objectAPI ObjectLayer, isSTS bool, m map[string]MappedPolicy) error {
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+	basePath := iamConfigPolicyDBUsersPrefix
+	if isSTS {
+		basePath = iamConfigPolicyDBSTSUsersPrefix
+	}
+	for item := range listIAMConfigItems(objectAPI, basePath, false, doneCh) {
+		if item.Err != nil {
+			return item.Err
+		}
+
+		policyFile := item.Item
+		userOrGroupName := strings.TrimSuffix(policyFile, ".json")
+		err := loadMappedPolicy(objectAPI, userOrGroupName, isSTS, m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 // LoadPolicy - reloads a specific canned policy from backend disks or etcd.
@@ -71,9 +332,8 @@ func (sys *IAMSys) LoadPolicy(objAPI ObjectLayer, policyName string) error {
 	sys.Lock()
 	defer sys.Unlock()
 
-	prefix := iamConfigPoliciesPrefix
 	if globalEtcdClient == nil {
-		return reloadPolicy(context.Background(), objAPI, prefix, policyName, sys.iamCannedPolicyMap)
+		return loadPolicyDoc(objAPI, policyName, sys.iamPolicyDocsMap)
 	}
 
 	// When etcd is set, we use watch APIs so this code is not needed.
@@ -81,7 +341,7 @@ func (sys *IAMSys) LoadPolicy(objAPI ObjectLayer, policyName string) error {
 }
 
 // LoadUser - reloads a specific user from backend disks or etcd.
-func (sys *IAMSys) LoadUser(objAPI ObjectLayer, accessKey string, temp bool) error {
+func (sys *IAMSys) LoadUser(objAPI ObjectLayer, accessKey string, isSTS bool) error {
 	if objAPI == nil {
 		return errInvalidArgument
 	}
@@ -89,13 +349,15 @@ func (sys *IAMSys) LoadUser(objAPI ObjectLayer, accessKey string, temp bool) err
 	sys.Lock()
 	defer sys.Unlock()
 
-	prefix := iamConfigUsersPrefix
-	if temp {
-		prefix = iamConfigSTSPrefix
-	}
-
 	if globalEtcdClient == nil {
-		return reloadUser(context.Background(), objAPI, prefix, accessKey, sys.iamUsersMap, sys.iamPolicyMap)
+		err := loadUser(objAPI, accessKey, isSTS, sys.iamUsersMap)
+		if err != nil {
+			return err
+		}
+		err = loadMappedPolicy(objAPI, accessKey, isSTS, sys.iamUserPolicyMap)
+		if err != nil {
+			return err
+		}
 	}
 	// When etcd is set, we use watch APIs so this code is not needed.
 	return nil
@@ -109,6 +371,7 @@ func (sys *IAMSys) Load(objAPI ObjectLayer) error {
 	return sys.refresh(objAPI)
 }
 
+// sys.RLock is held by caller.
 func (sys *IAMSys) reloadFromEvent(event *etcd.Event) {
 	eventCreate := event.IsModify() || event.IsCreate()
 	eventDelete := event.Type == etcd.EventTypeDelete
@@ -126,18 +389,15 @@ func (sys *IAMSys) reloadFromEvent(event *etcd.Event) {
 		case usersPrefix:
 			accessKey := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigUsersPrefix))
-			reloadEtcdUser(ctx, iamConfigUsersPrefix, accessKey,
-				sys.iamUsersMap, sys.iamPolicyMap)
+			loadEtcdUser(ctx, accessKey, false, sys.iamUsersMap)
 		case stsPrefix:
 			accessKey := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigSTSPrefix))
-			reloadEtcdUser(ctx, iamConfigSTSPrefix, accessKey,
-				sys.iamUsersMap, sys.iamPolicyMap)
+			loadEtcdUser(ctx, accessKey, true, sys.iamUsersMap)
 		case policyPrefix:
 			policyName := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigPoliciesPrefix))
-			reloadEtcdPolicy(ctx, iamConfigPoliciesPrefix,
-				policyName, sys.iamCannedPolicyMap)
+			loadEtcdPolicy(ctx, policyName, sys.iamPolicyDocsMap)
 		}
 	case eventDelete:
 		switch {
@@ -145,16 +405,14 @@ func (sys *IAMSys) reloadFromEvent(event *etcd.Event) {
 			accessKey := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigUsersPrefix))
 			delete(sys.iamUsersMap, accessKey)
-			delete(sys.iamPolicyMap, accessKey)
 		case stsPrefix:
 			accessKey := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigSTSPrefix))
 			delete(sys.iamUsersMap, accessKey)
-			delete(sys.iamPolicyMap, accessKey)
 		case policyPrefix:
 			policyName := path.Dir(strings.TrimPrefix(string(event.Kv.Key),
 				iamConfigPoliciesPrefix))
-			delete(sys.iamCannedPolicyMap, policyName)
+			delete(sys.iamPolicyDocsMap, policyName)
 		}
 	}
 }
@@ -208,10 +466,285 @@ func (sys *IAMSys) watchIAMDisk(objAPI ObjectLayer) {
 	go watchDisk()
 }
 
+// Migrate users directory in a single scan.
+//
+// 1. Migrate user policy from:
+//
+// `iamConfigUsersPrefix + "<username>/policy.json"`
+//
+// to:
+//
+// `iamConfigPolicyDBUsersPrefix + "<username>.json"`.
+//
+// 2. Add versioning to the policy json file in the new
+// location.
+//
+// 3. Migrate user identity json file to include version info.
+func migrateUsersConfigToV1(objAPI ObjectLayer, isSTS bool) error {
+	basePrefix := iamConfigUsersPrefix
+	if isSTS {
+		basePrefix = iamConfigSTSPrefix
+	}
+
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+	for item := range listIAMConfigItems(objAPI, basePrefix, true, doneCh) {
+		if item.Err != nil {
+			return item.Err
+		}
+
+		user := item.Item
+
+		{
+			// 1. check if there is policy file in old location.
+			oldPolicyPath := pathJoin(basePrefix, user, iamPolicyFile)
+			var policyName string
+			if err := loadIAMConfigItem(objAPI, &policyName, oldPolicyPath); err != nil {
+				switch err {
+				case errConfigNotFound:
+					// This case means it is already
+					// migrated or there is no policy on
+					// user.
+				default:
+					// File may be corrupt or network error
+				}
+
+				// Nothing to do on the policy file,
+				// so move on to check the id file.
+				goto next
+			}
+
+			// 2. copy policy file to new location.
+			mp := newMappedPolicy(policyName)
+			path := getMappedPolicyPath(user, isSTS)
+			if err := saveIAMConfigItem(objAPI, mp, path); err != nil {
+				return err
+			}
+
+			// 3. delete policy file from old
+			// location. Ignore error.
+			deleteConfig(context.Background(), objAPI, oldPolicyPath)
+		}
+	next:
+		// 4. check if user identity has old format.
+		identityPath := pathJoin(basePrefix, user, iamIdentityFile)
+		var cred auth.Credentials
+		if err := loadIAMConfigItem(objAPI, &cred, identityPath); err != nil {
+			switch err.(type) {
+			case ObjectNotFound:
+				// This should not happen.
+			default:
+				// File may be corrupt or network error
+			}
+			continue
+		}
+
+		// If the file is already in the new format,
+		// then the parsed auth.Credentials will have
+		// the zero value for the struct.
+		var zeroCred auth.Credentials
+		if cred == zeroCred {
+			// nothing to do
+			continue
+		}
+
+		// Found a id file in old format. Copy value
+		// into new format and save it.
+		u := newUserIdentity(cred)
+		if err := saveIAMConfigItem(objAPI, u, identityPath); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+
+		// Nothing to delete as identity file location
+		// has not changed.
+	}
+	return nil
+}
+
+// migrateUsersConfigEtcd - same as migrateUsersConfig but on etcd.
+func migrateUsersConfigEtcdToV1(isSTS bool) error {
+	basePrefix := iamConfigUsersPrefix
+	if isSTS {
+		basePrefix = iamConfigSTSPrefix
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
+	defer cancel()
+	r, err := globalEtcdClient.Get(ctx, basePrefix, etcd.WithPrefix(), etcd.WithKeysOnly())
+	if err != nil {
+		return err
+	}
+
+	users := etcdKvsToSet(basePrefix, r.Kvs)
+	for _, user := range users.ToSlice() {
+		{
+			// 1. check if there is a policy file in the old loc.
+			oldPolicyPath := pathJoin(basePrefix, user, iamPolicyFile)
+			var policyName string
+			err := loadIAMConfigItemEtcd(ctx, &policyName, oldPolicyPath)
+			if err != nil {
+				switch err {
+				case errConfigNotFound:
+					// No mapped policy or already migrated.
+				default:
+					// corrupt data/read error, etc
+				}
+				goto next
+			}
+
+			// 2. copy policy to new loc.
+			mp := newMappedPolicy(policyName)
+			path := getMappedPolicyPath(user, isSTS)
+			if err := saveIAMConfigItemEtcd(ctx, mp, path); err != nil {
+				return err
+			}
+
+			// 3. delete policy file in old loc.
+			deleteConfigEtcd(ctx, globalEtcdClient, path)
+		}
+
+	next:
+		// 4. check if user identity has old format.
+		identityPath := pathJoin(basePrefix, user, iamIdentityFile)
+		var cred auth.Credentials
+		if err := loadIAMConfigItemEtcd(ctx, &cred, identityPath); err != nil {
+			switch err {
+			case errConfigNotFound:
+				// This case should not happen.
+			default:
+				// corrupt file or read error
+			}
+			continue
+		}
+
+		// If the file is already in the new format,
+		// then the parsed auth.Credentials will have
+		// the zero value for the struct.
+		var zeroCred auth.Credentials
+		if cred == zeroCred {
+			// nothing to do
+			continue
+		}
+
+		// Found a id file in old format. Copy value
+		// into new format and save it.
+		u := newUserIdentity(cred)
+		if err := saveIAMConfigItemEtcd(ctx, u, identityPath); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+
+		// Nothing to delete as identity file location
+		// has not changed.
+	}
+	return nil
+
+}
+
+func migrateToV1(objAPI ObjectLayer) error {
+	currentIAMFormat := getCurrentIAMFormat()
+	path := getIAMFormatFilePath()
+
+	if globalEtcdClient == nil {
+		var iamFmt iamFormat
+		if err := loadIAMConfigItem(objAPI, &iamFmt, path); err != nil {
+			switch err {
+			case errConfigNotFound:
+				// Need to migrate to V1.
+			default:
+				return errors.New("corrupt IAM format file")
+			}
+		} else {
+			if iamFmt.Version >= currentIAMFormat.Version {
+				// Already migrated to V1 or higher!
+				return nil
+			}
+			// This case should not happen
+			// (i.e. Version is 0 or negative.)
+			return errors.New("got an invalid IAM format version")
+		}
+
+		// Migrate long-term users
+		if err := migrateUsersConfigToV1(objAPI, false); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+		// Migrate STS users
+		if err := migrateUsersConfigToV1(objAPI, true); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+		// Save iam version file.
+		if err := saveIAMConfigItem(objAPI, currentIAMFormat, path); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+	} else {
+		var iamFmt iamFormat
+		if err := loadIAMConfigItemEtcd(context.Background(), &iamFmt, path); err != nil {
+			switch err {
+			case errConfigNotFound:
+				// Need to migrate to V1.
+			default:
+				return errors.New("corrupt IAM format file")
+			}
+		} else {
+			if iamFmt.Version >= currentIAMFormat.Version {
+				// Already migrated to V1 of higher!
+				return nil
+			}
+			// This case should not happen
+			// (i.e. Version is 0 or negative.)
+			return errors.New("got an invalid IAM format version")
+
+		}
+
+		// Migrate long-term users
+		if err := migrateUsersConfigEtcdToV1(false); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+		// Migrate STS users
+		if err := migrateUsersConfigEtcdToV1(true); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+		// Save iam version file.
+		if err := saveIAMConfigItemEtcd(context.Background(), currentIAMFormat, path); err != nil {
+			logger.LogIf(context.Background(), err)
+			return err
+		}
+	}
+	return nil
+}
+
+// Perform IAM configuration migration.
+func doIAMConfigMigration(objAPI ObjectLayer) error {
+	// Take IAM configuration migration lock
+	lockPath := iamConfigPrefix + "/migration.lock"
+	objLock := globalNSMutex.NewNSLock(context.Background(), minioMetaBucket, lockPath)
+	if err := objLock.GetLock(globalOperationTimeout); err != nil {
+		return err
+	}
+	defer objLock.Unlock()
+
+	if err := migrateToV1(objAPI); err != nil {
+		return err
+	}
+	// Add future IAM migrations here.
+	return nil
+}
+
 // Init - initializes config system from iam.json
 func (sys *IAMSys) Init(objAPI ObjectLayer) error {
 	if objAPI == nil {
 		return errInvalidArgument
+	}
+
+	// Migrate IAM configuration
+	if err := doIAMConfigMigration(objAPI); err != nil {
+		return err
 	}
 
 	if globalEtcdClient != nil {
@@ -258,13 +791,12 @@ func (sys *IAMSys) DeletePolicy(policyName string) error {
 	}
 
 	var err error
-	pFile := pathJoin(iamConfigPoliciesPrefix, policyName, iamPolicyFile)
+	pFile := getPolicyDocPath(policyName)
 	if globalEtcdClient != nil {
 		err = deleteConfigEtcd(context.Background(), globalEtcdClient, pFile)
 	} else {
 		err = deleteConfig(context.Background(), objectAPI, pFile)
 	}
-
 	switch err.(type) {
 	case ObjectNotFound:
 		// Ignore error if policy is already deleted.
@@ -274,7 +806,7 @@ func (sys *IAMSys) DeletePolicy(policyName string) error {
 	sys.Lock()
 	defer sys.Unlock()
 
-	delete(sys.iamCannedPolicyMap, policyName)
+	delete(sys.iamPolicyDocsMap, policyName)
 	return err
 }
 
@@ -285,23 +817,23 @@ func (sys *IAMSys) ListPolicies() (map[string][]byte, error) {
 		return nil, errServerNotInitialized
 	}
 
-	var cannedPolicyMap = make(map[string][]byte)
+	var policyDocsMap = make(map[string][]byte)
 
 	sys.RLock()
 	defer sys.RUnlock()
 
-	for k, v := range sys.iamCannedPolicyMap {
+	for k, v := range sys.iamPolicyDocsMap {
 		data, err := json.Marshal(v)
 		if err != nil {
 			return nil, err
 		}
-		cannedPolicyMap[k] = data
+		policyDocsMap[k] = data
 	}
 
-	return cannedPolicyMap, nil
+	return policyDocsMap, nil
 }
 
-// SetPolicy - sets a new canned policy.
+// SetPolicy - sets a new name policy.
 func (sys *IAMSys) SetPolicy(policyName string, p iampolicy.Policy) error {
 	objectAPI := newObjectLayerFn()
 	if objectAPI == nil {
@@ -312,16 +844,12 @@ func (sys *IAMSys) SetPolicy(policyName string, p iampolicy.Policy) error {
 		return errInvalidArgument
 	}
 
-	configFile := pathJoin(iamConfigPoliciesPrefix, policyName, iamPolicyFile)
-	data, err := json.Marshal(p)
-	if err != nil {
-		return err
-	}
-
+	path := getPolicyDocPath(policyName)
+	var err error
 	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
+		err = saveIAMConfigItemEtcd(context.Background(), p, path)
 	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
+		err = saveIAMConfigItem(objectAPI, p, path)
 	}
 	if err != nil {
 		return err
@@ -330,49 +858,12 @@ func (sys *IAMSys) SetPolicy(policyName string, p iampolicy.Policy) error {
 	sys.Lock()
 	defer sys.Unlock()
 
-	sys.iamCannedPolicyMap[policyName] = p
+	sys.iamPolicyDocsMap[policyName] = p
 
 	return nil
 }
 
-// SetUserPolicy - sets policy to given user name.
-func (sys *IAMSys) SetUserPolicy(accessKey, policyName string) error {
-	objectAPI := newObjectLayerFn()
-	if objectAPI == nil {
-		return errServerNotInitialized
-	}
-
-	sys.Lock()
-	defer sys.Unlock()
-
-	if _, ok := sys.iamUsersMap[accessKey]; !ok {
-		return errNoSuchUser
-	}
-
-	if _, ok := sys.iamCannedPolicyMap[policyName]; !ok {
-		return errNoSuchPolicy
-	}
-
-	data, err := json.Marshal(policyName)
-	if err != nil {
-		return err
-	}
-
-	configFile := pathJoin(iamConfigUsersPrefix, accessKey, iamPolicyFile)
-	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
-	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
-	}
-	if err != nil {
-		return err
-	}
-
-	sys.iamPolicyMap[accessKey] = policyName
-	return nil
-}
-
-// DeleteUser - set user credentials.
+// DeleteUser - delete user (only for long-term users not STS users).
 func (sys *IAMSys) DeleteUser(accessKey string) error {
 	objectAPI := newObjectLayerFn()
 	if objectAPI == nil {
@@ -380,19 +871,18 @@ func (sys *IAMSys) DeleteUser(accessKey string) error {
 	}
 
 	var err error
-	pFile := pathJoin(iamConfigUsersPrefix, accessKey, iamPolicyFile)
-	iFile := pathJoin(iamConfigUsersPrefix, accessKey, iamIdentityFile)
+	mappingPath := getMappedPolicyPath(accessKey, false)
+	idPath := getUserIdentityPath(accessKey, false)
 	if globalEtcdClient != nil {
 		// It is okay to ignore errors when deleting policy.json for the user.
-		deleteConfigEtcd(context.Background(), globalEtcdClient, pFile)
-		err = deleteConfigEtcd(context.Background(), globalEtcdClient, iFile)
+		deleteConfigEtcd(context.Background(), globalEtcdClient, mappingPath)
+		err = deleteConfigEtcd(context.Background(), globalEtcdClient, idPath)
 	} else {
 		// It is okay to ignore errors when deleting policy.json for the user.
-		_ = deleteConfig(context.Background(), objectAPI, pFile)
-		err = deleteConfig(context.Background(), objectAPI, iFile)
+		_ = deleteConfig(context.Background(), objectAPI, mappingPath)
+		err = deleteConfig(context.Background(), objectAPI, idPath)
 	}
 
-	//
 	switch err.(type) {
 	case ObjectNotFound:
 		// ignore if user is already deleted.
@@ -403,7 +893,7 @@ func (sys *IAMSys) DeleteUser(accessKey string) error {
 	defer sys.Unlock()
 
 	delete(sys.iamUsersMap, accessKey)
-	delete(sys.iamPolicyMap, accessKey)
+	delete(sys.iamUserPolicyMap, accessKey)
 
 	return err
 }
@@ -422,72 +912,46 @@ func (sys *IAMSys) SetTempUser(accessKey string, cred auth.Credentials, policyNa
 	// temporary user which match with pre-configured canned
 	// policies for this server.
 	if globalPolicyOPA == nil && policyName != "" {
-		p, ok := sys.iamCannedPolicyMap[policyName]
+		p, ok := sys.iamPolicyDocsMap[policyName]
 		if !ok {
 			return errInvalidArgument
 		}
 		if p.IsEmpty() {
-			delete(sys.iamPolicyMap, accessKey)
+			delete(sys.iamUserPolicyMap, accessKey)
 			return nil
 		}
 
-		data, err := json.Marshal(policyName)
-		if err != nil {
-			return err
-		}
+		mp := newMappedPolicy(policyName)
 
-		configFile := pathJoin(iamConfigSTSPrefix, accessKey, iamPolicyFile)
+		mappingPath := getMappedPolicyPath(accessKey, true)
+		var err error
 		if globalEtcdClient != nil {
-			err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
+			err = saveIAMConfigItemEtcd(context.Background(), mp, mappingPath)
 		} else {
-			err = saveConfig(context.Background(), objectAPI, configFile, data)
+			err = saveIAMConfigItem(objectAPI, mp, mappingPath)
 		}
 		if err != nil {
 			return err
 		}
 
-		sys.iamPolicyMap[accessKey] = policyName
+		sys.iamUserPolicyMap[accessKey] = mp
 	}
 
-	configFile := pathJoin(iamConfigSTSPrefix, accessKey, iamIdentityFile)
-	data, err := json.Marshal(cred)
-	if err != nil {
-		return err
-	}
+	idPath := getUserIdentityPath(accessKey, true)
+	u := newUserIdentity(cred)
 
+	var err error
 	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
+		err = saveIAMConfigItemEtcd(context.Background(), u, idPath)
 	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
+		err = saveIAMConfigItem(objectAPI, u, idPath)
 	}
-
 	if err != nil {
 		return err
 	}
 
 	sys.iamUsersMap[accessKey] = cred
 	return nil
-}
-
-// GetUserPolicy - returns canned policy name associated with a user.
-func (sys *IAMSys) GetUserPolicy(accessKey string) (policyName string, err error) {
-	objectAPI := newObjectLayerFn()
-	if objectAPI == nil {
-		return "", errServerNotInitialized
-	}
-
-	sys.RLock()
-	defer sys.RUnlock()
-
-	if _, ok := sys.iamUsersMap[accessKey]; !ok {
-		return "", errNoSuchUser
-	}
-
-	if _, ok := sys.iamPolicyMap[accessKey]; !ok {
-		return "", errNoSuchUser
-	}
-
-	return sys.iamPolicyMap[accessKey], nil
 }
 
 // ListUsers - list all users.
@@ -504,7 +968,7 @@ func (sys *IAMSys) ListUsers() (map[string]madmin.UserInfo, error) {
 
 	for k, v := range sys.iamUsersMap {
 		users[k] = madmin.UserInfo{
-			PolicyName: sys.iamPolicyMap[k],
+			PolicyName: sys.iamUserPolicyMap[k].Policy,
 			Status:     madmin.AccountStatus(v.Status),
 		}
 	}
@@ -531,68 +995,58 @@ func (sys *IAMSys) SetUserStatus(accessKey string, status madmin.AccountStatus) 
 		return errNoSuchUser
 	}
 
-	uinfo := madmin.UserInfo{
-		SecretKey: cred.SecretKey,
-		Status:    status,
-	}
-
-	configFile := pathJoin(iamConfigUsersPrefix, accessKey, iamIdentityFile)
-	data, err := json.Marshal(uinfo)
-	if err != nil {
-		return err
-	}
-
-	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
-	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
-	}
-
-	if err != nil {
-		return err
-	}
-
-	sys.iamUsersMap[accessKey] = auth.Credentials{
+	uinfo := newUserIdentity(auth.Credentials{
 		AccessKey: accessKey,
-		SecretKey: uinfo.SecretKey,
-		Status:    string(uinfo.Status),
+		SecretKey: cred.SecretKey,
+		Status:    string(status),
+	})
+	idFile := getUserIdentityPath(accessKey, false)
+	var err error
+	if globalEtcdClient != nil {
+		err = saveIAMConfigItemEtcd(context.Background(), uinfo, idFile)
+	} else {
+		err = saveIAMConfigItem(objectAPI, uinfo, idFile)
+	}
+	if err != nil {
+		return err
 	}
 
+	sys.iamUsersMap[accessKey] = uinfo.Credentials
 	return nil
 }
 
-// SetUser - set user credentials.
+// SetUser - set user credentials and policy.
 func (sys *IAMSys) SetUser(accessKey string, uinfo madmin.UserInfo) error {
 	objectAPI := newObjectLayerFn()
 	if objectAPI == nil {
 		return errServerNotInitialized
 	}
 
-	configFile := pathJoin(iamConfigUsersPrefix, accessKey, iamIdentityFile)
-	data, err := json.Marshal(uinfo)
-	if err != nil {
-		return err
-	}
-
-	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
-	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
-	}
-
-	if err != nil {
-		return err
-	}
+	idFile := getUserIdentityPath(accessKey, false)
+	u := newUserIdentity(auth.Credentials{
+		AccessKey: accessKey,
+		SecretKey: uinfo.SecretKey,
+		Status:    string(uinfo.Status),
+	})
 
 	sys.Lock()
 	defer sys.Unlock()
 
-	sys.iamUsersMap[accessKey] = auth.Credentials{
-		AccessKey: accessKey,
-		SecretKey: uinfo.SecretKey,
-		Status:    string(uinfo.Status),
+	var err error
+	if globalEtcdClient != nil {
+		err = saveIAMConfigItemEtcd(context.Background(), u, idFile)
+	} else {
+		err = saveIAMConfigItem(objectAPI, u, idFile)
 	}
+	if err != nil {
+		return err
+	}
+	sys.iamUsersMap[accessKey] = u.Credentials
 
+	// Set policy if specified.
+	if uinfo.PolicyName != "" {
+		return sys.policyDBSet(objectAPI, accessKey, uinfo.PolicyName, false)
+	}
 	return nil
 }
 
@@ -611,33 +1065,21 @@ func (sys *IAMSys) SetUserSecretKey(accessKey string, secretKey string) error {
 		return errNoSuchUser
 	}
 
-	uinfo := madmin.UserInfo{
-		SecretKey: secretKey,
-		Status:    madmin.AccountStatus(cred.Status),
-	}
+	cred.SecretKey = secretKey
+	u := newUserIdentity(cred)
+	idFile := getUserIdentityPath(accessKey, false)
 
-	configFile := pathJoin(iamConfigUsersPrefix, accessKey, iamIdentityFile)
-	data, err := json.Marshal(uinfo)
-	if err != nil {
-		return err
-	}
-
+	var err error
 	if globalEtcdClient != nil {
-		err = saveConfigEtcd(context.Background(), globalEtcdClient, configFile, data)
+		err = saveIAMConfigItemEtcd(context.Background(), u, idFile)
 	} else {
-		err = saveConfig(context.Background(), objectAPI, configFile, data)
+		err = saveIAMConfigItem(objectAPI, u, idFile)
 	}
-
 	if err != nil {
 		return err
 	}
 
-	sys.iamUsersMap[accessKey] = auth.Credentials{
-		AccessKey: accessKey,
-		SecretKey: secretKey,
-		Status:    string(uinfo.Status),
-	}
-
+	sys.iamUsersMap[accessKey] = cred
 	return nil
 }
 
@@ -648,6 +1090,79 @@ func (sys *IAMSys) GetUser(accessKey string) (cred auth.Credentials, ok bool) {
 
 	cred, ok = sys.iamUsersMap[accessKey]
 	return cred, ok && cred.IsValid()
+}
+
+// PolicyDBSet - sets a policy for a user or group in the
+// PolicyDB. This function applies only long-term users. For STS
+// users, policy is set directly by called sys.policyDBSet().
+func (sys *IAMSys) PolicyDBSet(name, policy string) error {
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		return errServerNotInitialized
+	}
+
+	sys.Lock()
+	defer sys.Unlock()
+
+	return sys.policyDBSet(objectAPI, name, policy, false)
+}
+
+// policyDBSet - sets a policy for user in the policy db. Assumes that
+// caller has sys.Lock().
+func (sys *IAMSys) policyDBSet(objectAPI ObjectLayer, name, policy string, isSTS bool) error {
+	if name == "" || policy == "" {
+		return errInvalidArgument
+	}
+
+	if _, ok := sys.iamUsersMap[name]; !ok {
+		return errNoSuchUser
+	}
+
+	_, ok := sys.iamPolicyDocsMap[policy]
+	if !ok {
+		return errNoSuchPolicy
+	}
+
+	mp := newMappedPolicy(policy)
+	_, ok = sys.iamUsersMap[name]
+	if !ok {
+		return errNoSuchUser
+	}
+	var err error
+	mappingPath := getMappedPolicyPath(name, isSTS)
+	if globalEtcdClient != nil {
+		err = saveIAMConfigItemEtcd(context.Background(), mp, mappingPath)
+	} else {
+		err = saveIAMConfigItem(objectAPI, mp, mappingPath)
+	}
+	if err != nil {
+		return err
+	}
+	sys.iamUserPolicyMap[name] = mp
+	return nil
+}
+
+// PolicyDBGet - gets policy set on a user
+func (sys *IAMSys) PolicyDBGet(name string) (string, error) {
+	if name == "" {
+		return "", errInvalidArgument
+	}
+
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		return "", errServerNotInitialized
+	}
+
+	sys.RLock()
+	defer sys.RUnlock()
+
+	if _, ok := sys.iamUsersMap[name]; !ok {
+		return "", errNoSuchUser
+	}
+
+	policy := sys.iamUserPolicyMap[name]
+	// returned policy could be empty
+	return policy.Policy, nil
 }
 
 // IsAllowedSTS is meant for STS based temporary credentials,
@@ -669,11 +1184,12 @@ func (sys *IAMSys) IsAllowedSTS(args iampolicy.Args) bool {
 	defer sys.RUnlock()
 
 	// If policy is available for given user, check the policy.
-	name, ok := sys.iamPolicyMap[args.AccountName]
+	mp, ok := sys.iamUserPolicyMap[args.AccountName]
 	if !ok {
 		// No policy available reject.
 		return false
 	}
+	name := mp.Policy
 
 	if pnameStr != name {
 		// When claims has a policy, it should match the
@@ -687,7 +1203,7 @@ func (sys *IAMSys) IsAllowedSTS(args iampolicy.Args) bool {
 	if !ok {
 		// Sub policy not set, this is most common since subPolicy
 		// is optional, use the top level policy only.
-		p, ok := sys.iamCannedPolicyMap[pnameStr]
+		p, ok := sys.iamPolicyDocsMap[pnameStr]
 		return ok && p.IsAllowed(args)
 	}
 
@@ -712,7 +1228,7 @@ func (sys *IAMSys) IsAllowedSTS(args iampolicy.Args) bool {
 	}
 
 	// Sub policy is set and valid.
-	p, ok := sys.iamCannedPolicyMap[pnameStr]
+	p, ok := sys.iamPolicyDocsMap[pnameStr]
 	return ok && p.IsAllowed(args) && subPolicy.IsAllowed(args)
 }
 
@@ -732,12 +1248,13 @@ func (sys *IAMSys) IsAllowed(args iampolicy.Args) bool {
 	defer sys.RUnlock()
 
 	// If policy is available for given user, check the policy.
-	if name, found := sys.iamPolicyMap[args.AccountName]; found {
-		p, ok := sys.iamCannedPolicyMap[name]
+	if mp, found := sys.iamUserPolicyMap[args.AccountName]; found {
+		p, ok := sys.iamPolicyDocsMap[mp.Policy]
 		return ok && p.IsAllowed(args)
 	}
 
-	// As policy is not available and OPA is not configured, return the owner value.
+	// As policy is not available and OPA is not configured,
+	// return the owner value.
 	return args.IsOwner
 }
 
@@ -762,206 +1279,125 @@ func etcdKvsToSet(prefix string, kvs []*mvccpb.KeyValue) set.StringSet {
 	return users
 }
 
-// Similar to reloadUsers but updates users, policies maps from etcd server,
-func reloadEtcdUsers(prefix string, usersMap map[string]auth.Credentials, policyMap map[string]string) error {
-	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
-	defer cancel()
-	r, err := globalEtcdClient.Get(ctx, prefix, etcd.WithPrefix(), etcd.WithKeysOnly())
+func etcdKvsToSetPolicyDB(prefix string, kvs []*mvccpb.KeyValue) set.StringSet {
+	items := set.NewStringSet()
+	for _, kv := range kvs {
+		// Extract user item by stripping off prefix and then
+		// stripping of ".json" suffix.
+		//
+		// key := "config/iam/policydb/users/myuser1.json"
+		// prefix := "config/iam/policydb/users/"
+		// v := trimSuffix(trimPrefix(key, prefix), ".json")
+		key := string(kv.Key)
+		item := path.Clean(strings.TrimSuffix(strings.TrimPrefix(key, prefix), ".json"))
+		items.Add(item)
+	}
+	return items
+}
+
+func loadEtcdMappedPolicy(ctx context.Context, name string, isSTS bool,
+	m map[string]MappedPolicy) error {
+
+	var p MappedPolicy
+	err := loadIAMConfigItemEtcd(ctx, &p, getMappedPolicyPath(name, isSTS))
 	if err != nil {
 		return err
 	}
-	// No users are created yet.
-	if r.Count == 0 {
-		return nil
+	m[name] = p
+	return nil
+}
+
+func loadEtcdMappedPolicies(isSTS bool, m map[string]MappedPolicy) error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
+	defer cancel()
+	basePrefix := iamConfigPolicyDBUsersPrefix
+	if isSTS {
+		basePrefix = iamConfigPolicyDBSTSUsersPrefix
+	}
+	r, err := globalEtcdClient.Get(ctx, basePrefix, etcd.WithPrefix(), etcd.WithKeysOnly())
+	if err != nil {
+		return err
 	}
 
-	users := etcdKvsToSet(prefix, r.Kvs)
+	users := etcdKvsToSetPolicyDB(basePrefix, r.Kvs)
 
 	// Reload config and policies for all users.
 	for _, user := range users.ToSlice() {
-		if err = reloadEtcdUser(ctx, prefix, user, usersMap, policyMap); err != nil {
+		if err = loadEtcdMappedPolicy(ctx, user, isSTS, m); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func reloadEtcdPolicy(ctx context.Context, prefix string, policyName string,
-	cannedPolicyMap map[string]iampolicy.Policy) error {
-	pFile := pathJoin(prefix, policyName, iamPolicyFile)
-	pdata, err := readConfigEtcd(ctx, globalEtcdClient, pFile)
+func loadEtcdUser(ctx context.Context, user string, isSTS bool, m map[string]auth.Credentials) error {
+	var u UserIdentity
+	err := loadIAMConfigItemEtcd(ctx, &u, getUserIdentityPath(user, isSTS))
 	if err != nil {
 		return err
 	}
-	var p iampolicy.Policy
-	if err = json.Unmarshal(pdata, &p); err != nil {
-		return err
-	}
-	cannedPolicyMap[policyName] = p
-	return nil
-}
 
-func reloadEtcdPolicies(prefix string, cannedPolicyMap map[string]iampolicy.Policy) error {
-	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
-	defer cancel()
-	r, err := globalEtcdClient.Get(ctx, prefix, etcd.WithPrefix(), etcd.WithKeysOnly())
-	if err != nil {
-		return err
-	}
-	// No users are created yet.
-	if r.Count == 0 {
+	if u.Credentials.IsExpired() {
+		idPath := getUserIdentityPath(user, isSTS)
+		// Delete expired identity.
+		deleteConfigEtcd(ctx, globalEtcdClient, idPath)
+		deleteConfigEtcd(ctx, globalEtcdClient, getMappedPolicyPath(user, isSTS))
 		return nil
 	}
 
-	policies := etcdKvsToSet(prefix, r.Kvs)
+	m[user] = u.Credentials
+	return nil
+}
+
+func loadEtcdUsers(isSTS bool, m map[string]auth.Credentials) error {
+	basePrefix := iamConfigUsersPrefix
+	if isSTS {
+		basePrefix = iamConfigSTSPrefix
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
+	defer cancel()
+	r, err := globalEtcdClient.Get(ctx, basePrefix, etcd.WithPrefix(), etcd.WithKeysOnly())
+	if err != nil {
+		return err
+	}
+
+	users := etcdKvsToSet(basePrefix, r.Kvs)
+
+	// Reload config and policies for all users.
+	for _, user := range users.ToSlice() {
+		if err = loadEtcdUser(ctx, user, isSTS, m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func loadEtcdPolicy(ctx context.Context, policyName string, m map[string]iampolicy.Policy) error {
+	var p iampolicy.Policy
+	err := loadIAMConfigItemEtcd(ctx, &p, getPolicyDocPath(policyName))
+	if err != nil {
+		return err
+	}
+	m[policyName] = p
+	return nil
+}
+
+func loadEtcdPolicies(policyDocsMap map[string]iampolicy.Policy) error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultContextTimeout)
+	defer cancel()
+	r, err := globalEtcdClient.Get(ctx, iamConfigPoliciesPrefix, etcd.WithPrefix(), etcd.WithKeysOnly())
+	if err != nil {
+		return err
+	}
+
+	policies := etcdKvsToSet(iamConfigPoliciesPrefix, r.Kvs)
 
 	// Reload config and policies for all policys.
 	for _, policyName := range policies.ToSlice() {
-		if err = reloadEtcdPolicy(ctx, prefix, policyName, cannedPolicyMap); err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-func reloadPolicy(ctx context.Context, objectAPI ObjectLayer, prefix string,
-	policyName string, cannedPolicyMap map[string]iampolicy.Policy) error {
-	pFile := pathJoin(prefix, policyName, iamPolicyFile)
-	pdata, err := readConfig(context.Background(), objectAPI, pFile)
-	if err != nil {
-		return err
-	}
-	var p iampolicy.Policy
-	if err = json.Unmarshal(pdata, &p); err != nil {
-		return err
-	}
-	cannedPolicyMap[policyName] = p
-	return nil
-}
-
-func reloadPolicies(objectAPI ObjectLayer, prefix string, cannedPolicyMap map[string]iampolicy.Policy) error {
-	marker := ""
-	for {
-		var lo ListObjectsInfo
-		var err error
-		lo, err = objectAPI.ListObjects(context.Background(), minioMetaBucket, prefix, marker, "/", 1000)
+		err = loadEtcdPolicy(ctx, policyName, policyDocsMap)
 		if err != nil {
 			return err
-		}
-		marker = lo.NextMarker
-		for _, prefix := range lo.Prefixes {
-			if err = reloadPolicy(context.Background(), objectAPI, iamConfigPoliciesPrefix,
-				path.Base(prefix), cannedPolicyMap); err != nil {
-				return err
-			}
-		}
-		if !lo.IsTruncated {
-			break
-		}
-	}
-	return nil
-
-}
-
-func reloadEtcdUser(ctx context.Context, prefix string, accessKey string,
-	usersMap map[string]auth.Credentials, policyMap map[string]string) error {
-	idFile := pathJoin(prefix, accessKey, iamIdentityFile)
-	pFile := pathJoin(prefix, accessKey, iamPolicyFile)
-	cdata, cerr := readConfigEtcd(ctx, globalEtcdClient, idFile)
-	pdata, perr := readConfigEtcd(ctx, globalEtcdClient, pFile)
-	if cerr != nil && cerr != errConfigNotFound {
-		return cerr
-	}
-	if perr != nil && perr != errConfigNotFound {
-		return perr
-	}
-	if cerr == errConfigNotFound && perr == errConfigNotFound {
-		return nil
-	}
-	if cerr == nil {
-		var cred auth.Credentials
-		if err := json.Unmarshal(cdata, &cred); err != nil {
-			return err
-		}
-		cred.AccessKey = path.Base(accessKey)
-		if cred.IsExpired() {
-			// Delete expired identity.
-			deleteConfigEtcd(ctx, globalEtcdClient, idFile)
-			// Delete expired identity policy.
-			deleteConfigEtcd(ctx, globalEtcdClient, pFile)
-			return nil
-		}
-		usersMap[cred.AccessKey] = cred
-	}
-	if perr == nil {
-		var policyName string
-		if err := json.Unmarshal(pdata, &policyName); err != nil {
-			return err
-		}
-		policyMap[path.Base(accessKey)] = policyName
-	}
-	return nil
-}
-
-func reloadUser(ctx context.Context, objectAPI ObjectLayer, prefix string, accessKey string,
-	usersMap map[string]auth.Credentials, policyMap map[string]string) error {
-	idFile := pathJoin(prefix, accessKey, iamIdentityFile)
-	pFile := pathJoin(prefix, accessKey, iamPolicyFile)
-	cdata, cerr := readConfig(ctx, objectAPI, idFile)
-	pdata, perr := readConfig(ctx, objectAPI, pFile)
-	if cerr != nil && cerr != errConfigNotFound {
-		return cerr
-	}
-	if perr != nil && perr != errConfigNotFound {
-		return perr
-	}
-	if cerr == errConfigNotFound && perr == errConfigNotFound {
-		return nil
-	}
-	if cerr == nil {
-		var cred auth.Credentials
-		if err := json.Unmarshal(cdata, &cred); err != nil {
-			return err
-		}
-		cred.AccessKey = path.Base(accessKey)
-		if cred.IsExpired() {
-			// Delete expired identity.
-			objectAPI.DeleteObject(context.Background(), minioMetaBucket, idFile)
-			// Delete expired identity policy.
-			objectAPI.DeleteObject(context.Background(), minioMetaBucket, pFile)
-			return nil
-		}
-		usersMap[cred.AccessKey] = cred
-	}
-	if perr == nil {
-		var policyName string
-		if err := json.Unmarshal(pdata, &policyName); err != nil {
-			return err
-		}
-		policyMap[path.Base(accessKey)] = policyName
-	}
-	return nil
-}
-
-// reloadUsers reads an updates users, policies from object layer into user and policy maps.
-func reloadUsers(objectAPI ObjectLayer, prefix string, usersMap map[string]auth.Credentials, policyMap map[string]string) error {
-	marker := ""
-	for {
-		var lo ListObjectsInfo
-		var err error
-		lo, err = objectAPI.ListObjects(context.Background(), minioMetaBucket, prefix, marker, "/", 1000)
-		if err != nil {
-			return err
-		}
-		marker = lo.NextMarker
-		for _, prefix := range lo.Prefixes {
-			// Prefix is empty because prefix is already part of the List output.
-			if err = reloadUser(context.Background(), objectAPI, "", prefix, usersMap, policyMap); err != nil {
-				return err
-			}
-		}
-		if !lo.IsTruncated {
-			break
 		}
 	}
 	return nil
@@ -985,28 +1421,36 @@ func setDefaultCannedPolicies(policies map[string]iampolicy.Policy) {
 
 func (sys *IAMSys) refreshEtcd() error {
 	iamUsersMap := make(map[string]auth.Credentials)
-	iamPolicyMap := make(map[string]string)
-	iamCannedPolicyMap := make(map[string]iampolicy.Policy)
+	iamPolicyDocsMap := make(map[string]iampolicy.Policy)
+	iamUserPolicyMap := make(map[string]MappedPolicy)
 
-	if err := reloadEtcdPolicies(iamConfigPoliciesPrefix, iamCannedPolicyMap); err != nil {
+	if err := loadEtcdPolicies(iamPolicyDocsMap); err != nil {
 		return err
 	}
-	if err := reloadEtcdUsers(iamConfigUsersPrefix, iamUsersMap, iamPolicyMap); err != nil {
+	if err := loadEtcdUsers(false, iamUsersMap); err != nil {
 		return err
 	}
-	if err := reloadEtcdUsers(iamConfigSTSPrefix, iamUsersMap, iamPolicyMap); err != nil {
+	// load STS temp users into the same map
+	if err := loadEtcdUsers(true, iamUsersMap); err != nil {
+		return err
+	}
+	if err := loadEtcdMappedPolicies(false, iamUserPolicyMap); err != nil {
+		return err
+	}
+	// load STS users policy mapping into the same map
+	if err := loadEtcdMappedPolicies(true, iamUserPolicyMap); err != nil {
 		return err
 	}
 
 	// Sets default canned policies, if none are set.
-	setDefaultCannedPolicies(iamCannedPolicyMap)
+	setDefaultCannedPolicies(iamPolicyDocsMap)
 
 	sys.Lock()
 	defer sys.Unlock()
 
 	sys.iamUsersMap = iamUsersMap
-	sys.iamPolicyMap = iamPolicyMap
-	sys.iamCannedPolicyMap = iamCannedPolicyMap
+	sys.iamUserPolicyMap = iamUserPolicyMap
+	sys.iamPolicyDocsMap = iamPolicyDocsMap
 
 	return nil
 }
@@ -1014,28 +1458,37 @@ func (sys *IAMSys) refreshEtcd() error {
 // Refresh IAMSys.
 func (sys *IAMSys) refresh(objAPI ObjectLayer) error {
 	iamUsersMap := make(map[string]auth.Credentials)
-	iamPolicyMap := make(map[string]string)
-	iamCannedPolicyMap := make(map[string]iampolicy.Policy)
+	iamPolicyDocsMap := make(map[string]iampolicy.Policy)
+	iamUserPolicyMap := make(map[string]MappedPolicy)
 
-	if err := reloadPolicies(objAPI, iamConfigPoliciesPrefix, iamCannedPolicyMap); err != nil {
+	if err := loadPolicyDocs(objAPI, iamPolicyDocsMap); err != nil {
 		return err
 	}
-	if err := reloadUsers(objAPI, iamConfigUsersPrefix, iamUsersMap, iamPolicyMap); err != nil {
+	if err := loadUsers(objAPI, false, iamUsersMap); err != nil {
 		return err
 	}
-	if err := reloadUsers(objAPI, iamConfigSTSPrefix, iamUsersMap, iamPolicyMap); err != nil {
+	// load STS temp users into the same map
+	if err := loadUsers(objAPI, true, iamUsersMap); err != nil {
+		return err
+	}
+
+	if err := loadMappedPolicies(objAPI, false, iamUserPolicyMap); err != nil {
+		return err
+	}
+	// load STS policy mappings into the same map
+	if err := loadMappedPolicies(objAPI, true, iamUserPolicyMap); err != nil {
 		return err
 	}
 
 	// Sets default canned policies, if none are set.
-	setDefaultCannedPolicies(iamCannedPolicyMap)
+	setDefaultCannedPolicies(iamPolicyDocsMap)
 
 	sys.Lock()
 	defer sys.Unlock()
 
 	sys.iamUsersMap = iamUsersMap
-	sys.iamPolicyMap = iamPolicyMap
-	sys.iamCannedPolicyMap = iamCannedPolicyMap
+	sys.iamPolicyDocsMap = iamPolicyDocsMap
+	sys.iamUserPolicyMap = iamUserPolicyMap
 
 	return nil
 }
@@ -1043,8 +1496,8 @@ func (sys *IAMSys) refresh(objAPI ObjectLayer) error {
 // NewIAMSys - creates new config system object.
 func NewIAMSys() *IAMSys {
 	return &IAMSys{
-		iamUsersMap:        make(map[string]auth.Credentials),
-		iamPolicyMap:       make(map[string]string),
-		iamCannedPolicyMap: make(map[string]iampolicy.Policy),
+		iamUsersMap:      make(map[string]auth.Credentials),
+		iamPolicyDocsMap: make(map[string]iampolicy.Policy),
+		iamUserPolicyMap: make(map[string]MappedPolicy),
 	}
 }

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -601,7 +601,7 @@ func migrateUsersConfigEtcdToV1(isSTS bool) error {
 			}
 
 			// 3. delete policy file in old loc.
-			deleteConfigEtcd(ctx, globalEtcdClient, path)
+			deleteConfigEtcd(ctx, globalEtcdClient, oldPolicyPath)
 		}
 
 	next:

--- a/cmd/sts-handlers.go
+++ b/cmd/sts-handlers.go
@@ -183,7 +183,7 @@ func (sts *stsAPIHandlers) AssumeRole(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	policyName, err := globalIAMSys.GetUserPolicy(user.AccessKey)
+	policyName, err := globalIAMSys.PolicyDBGet(user.AccessKey)
 	if err != nil {
 		logger.LogIf(ctx, err)
 		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))

